{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, cbind, colMeans, colnames,\n",
      "    colSums, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, lengths, Map, mapply, match,\n",
      "    mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    apply\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape’\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand, rename\n",
      "\n",
      "\n",
      "Attaching package: ‘plotly’\n",
      "\n",
      "The following object is masked from ‘package:reshape’:\n",
      "\n",
      "    rename\n",
      "\n",
      "The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    last_plot\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "The following object is masked from ‘package:graphics’:\n",
      "\n",
      "    layout\n",
      "\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "The following objects are masked from ‘package:plotly’:\n",
      "\n",
      "    arrange, mutate, rename, summarise\n",
      "\n",
      "The following objects are masked from ‘package:reshape’:\n",
      "\n",
      "    rename, round_any\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    desc\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "Loading required package: futile.logger\n"
     ]
    }
   ],
   "source": [
    "library(\"DESeq2\")\n",
    "library(\"pheatmap\")\n",
    "library(\"ggplot2\")\n",
    "library(\"reshape\")\n",
    "library(\"plotly\")\n",
    "library(\"grid\")\n",
    "library(\"gridExtra\")\n",
    "library(\"plyr\")\n",
    "library(\"VennDiagram\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Multiple plot function\n",
    "#\n",
    "# ggplot objects can be passed in ..., or to plotlist (as a list of ggplot objects)\n",
    "# - cols:   Number of columns in layout\n",
    "# - layout: A matrix specifying the layout. If present, 'cols' is ignored.\n",
    "#\n",
    "# If the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE),\n",
    "# then plot 1 will go in the upper left, 2 will go in the upper right, and\n",
    "# 3 will go all the way across the bottom.\n",
    "#\n",
    "multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {\n",
    "  library(grid)\n",
    "\n",
    "  # Make a list from the ... arguments and plotlist\n",
    "  plots <- c(list(...), plotlist)\n",
    "\n",
    "  numPlots = length(plots)\n",
    "\n",
    "  # If layout is NULL, then use 'cols' to determine layout\n",
    "  if (is.null(layout)) {\n",
    "    # Make the panel\n",
    "    # ncol: Number of columns of plots\n",
    "    # nrow: Number of rows needed, calculated from # of cols\n",
    "    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),\n",
    "                    ncol = cols, nrow = ceiling(numPlots/cols))\n",
    "  }\n",
    "\n",
    " if (numPlots==1) {\n",
    "    print(plots[[1]])\n",
    "\n",
    "  } else {\n",
    "    # Set up the page\n",
    "    grid.newpage()\n",
    "    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))\n",
    "\n",
    "    # Make each plot, in the correct location\n",
    "    for (i in 1:numPlots) {\n",
    "      # Get the i,j matrix positions of the regions that contain this subplot\n",
    "      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))\n",
    "\n",
    "      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,\n",
    "                                      layout.pos.col = matchidx$col))\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import data\n",
    "\n",
    "Here, I import the un-normalized read counts from StringTie output as a count matrix for DESeq2.\n",
    "\n",
    "First, import the gene counts as a matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>R1C24A</th><th scope=col>R1C24B</th><th scope=col>R1C24C</th><th scope=col>R1C24D</th><th scope=col>R1C48A</th><th scope=col>R1C48B</th><th scope=col>R1C48C</th><th scope=col>R1C48D</th><th scope=col>R1C72A</th><th scope=col>R1C72B</th><th scope=col>⋯</th><th scope=col>S2Q24C</th><th scope=col>S2Q24D</th><th scope=col>S2Q48A</th><th scope=col>S2Q48B</th><th scope=col>S2Q48C</th><th scope=col>S2Q48D</th><th scope=col>S2Q72A</th><th scope=col>S2Q72B</th><th scope=col>S2Q72C</th><th scope=col>S2Q72D</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>  26</td><td>  24</td><td>  22</td><td>  22</td><td>  21</td><td>  21</td><td>  27</td><td>  25</td><td>  25</td><td>  36</td><td>⋯   </td><td>  17</td><td>  12</td><td>  16</td><td>  13</td><td>  19</td><td>  20</td><td>   0</td><td>  12</td><td>  19</td><td>  15</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1849</td><td>1981</td><td>1858</td><td>1955</td><td>2183</td><td>2243</td><td>2156</td><td>2372</td><td>3410</td><td>3571</td><td>⋯   </td><td>1986</td><td>1981</td><td>2208</td><td>2191</td><td>2095</td><td>2229</td><td>2013</td><td>2226</td><td>2125</td><td>2244</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>  17</td><td>  30</td><td>   0</td><td>  27</td><td>  12</td><td>  21</td><td>   1</td><td>  17</td><td>  10</td><td>  44</td><td>⋯   </td><td>   3</td><td>  10</td><td>  20</td><td>   0</td><td>  10</td><td>  10</td><td>   2</td><td>   9</td><td>   7</td><td>  13</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>  78</td><td>  32</td><td>  52</td><td>  42</td><td>  26</td><td>  25</td><td>  55</td><td>  15</td><td>  48</td><td>  27</td><td>⋯   </td><td>  37</td><td>   0</td><td>  29</td><td>  55</td><td>  27</td><td>  23</td><td>  36</td><td>  32</td><td>  11</td><td>  33</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>  55</td><td>  54</td><td>  36</td><td>  80</td><td>  59</td><td>  44</td><td>  52</td><td>  67</td><td> 145</td><td> 119</td><td>⋯   </td><td> 158</td><td> 170</td><td> 199</td><td> 177</td><td> 182</td><td> 182</td><td> 155</td><td> 185</td><td> 176</td><td> 182</td></tr>\n",
       "\t<tr><th scope=row>6</th><td> 426</td><td> 357</td><td> 515</td><td> 597</td><td> 631</td><td> 588</td><td> 590</td><td> 604</td><td> 881</td><td> 576</td><td>⋯   </td><td> 690</td><td> 495</td><td> 638</td><td> 615</td><td> 593</td><td> 631</td><td> 672</td><td> 529</td><td> 672</td><td> 621</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & R1C24A & R1C24B & R1C24C & R1C24D & R1C48A & R1C48B & R1C48C & R1C48D & R1C72A & R1C72B & ⋯ & S2Q24C & S2Q24D & S2Q48A & S2Q48B & S2Q48C & S2Q48D & S2Q72A & S2Q72B & S2Q72C & S2Q72D\\\\\n",
       "\\hline\n",
       "\t1 &   26 &   24 &   22 &   22 &   21 &   21 &   27 &   25 &   25 &   36 & ⋯    &   17 &   12 &   16 &   13 &   19 &   20 &    0 &   12 &   19 &   15\\\\\n",
       "\t2 & 1849 & 1981 & 1858 & 1955 & 2183 & 2243 & 2156 & 2372 & 3410 & 3571 & ⋯    & 1986 & 1981 & 2208 & 2191 & 2095 & 2229 & 2013 & 2226 & 2125 & 2244\\\\\n",
       "\t3 &   17 &   30 &    0 &   27 &   12 &   21 &    1 &   17 &   10 &   44 & ⋯    &    3 &   10 &   20 &    0 &   10 &   10 &    2 &    9 &    7 &   13\\\\\n",
       "\t4 &   78 &   32 &   52 &   42 &   26 &   25 &   55 &   15 &   48 &   27 & ⋯    &   37 &    0 &   29 &   55 &   27 &   23 &   36 &   32 &   11 &   33\\\\\n",
       "\t5 &   55 &   54 &   36 &   80 &   59 &   44 &   52 &   67 &  145 &  119 & ⋯    &  158 &  170 &  199 &  177 &  182 &  182 &  155 &  185 &  176 &  182\\\\\n",
       "\t6 &  426 &  357 &  515 &  597 &  631 &  588 &  590 &  604 &  881 &  576 & ⋯    &  690 &  495 &  638 &  615 &  593 &  631 &  672 &  529 &  672 &  621\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | R1C24A | R1C24B | R1C24C | R1C24D | R1C48A | R1C48B | R1C48C | R1C48D | R1C72A | R1C72B | ⋯ | S2Q24C | S2Q24D | S2Q48A | S2Q48B | S2Q48C | S2Q48D | S2Q72A | S2Q72B | S2Q72C | S2Q72D | \n",
       "|---|---|---|---|---|---|\n",
       "| 1 |   26 |   24 |   22 |   22 |   21 |   21 |   27 |   25 |   25 |   36 | ⋯    |   17 |   12 |   16 |   13 |   19 |   20 |    0 |   12 |   19 |   15 | \n",
       "| 2 | 1849 | 1981 | 1858 | 1955 | 2183 | 2243 | 2156 | 2372 | 3410 | 3571 | ⋯    | 1986 | 1981 | 2208 | 2191 | 2095 | 2229 | 2013 | 2226 | 2125 | 2244 | \n",
       "| 3 |   17 |   30 |    0 |   27 |   12 |   21 |    1 |   17 |   10 |   44 | ⋯    |    3 |   10 |   20 |    0 |   10 |   10 |    2 |    9 |    7 |   13 | \n",
       "| 4 |   78 |   32 |   52 |   42 |   26 |   25 |   55 |   15 |   48 |   27 | ⋯    |   37 |    0 |   29 |   55 |   27 |   23 |   36 |   32 |   11 |   33 | \n",
       "| 5 |   55 |   54 |   36 |   80 |   59 |   44 |   52 |   67 |  145 |  119 | ⋯    |  158 |  170 |  199 |  177 |  182 |  182 |  155 |  185 |  176 |  182 | \n",
       "| 6 |  426 |  357 |  515 |  597 |  631 |  588 |  590 |  604 |  881 |  576 | ⋯    |  690 |  495 |  638 |  615 |  593 |  631 |  672 |  529 |  672 |  621 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  R1C24A R1C24B R1C24C R1C24D R1C48A R1C48B R1C48C R1C48D R1C72A R1C72B ⋯\n",
       "1   26     24     22     22     21     21     27     25     25     36   ⋯\n",
       "2 1849   1981   1858   1955   2183   2243   2156   2372   3410   3571   ⋯\n",
       "3   17     30      0     27     12     21      1     17     10     44   ⋯\n",
       "4   78     32     52     42     26     25     55     15     48     27   ⋯\n",
       "5   55     54     36     80     59     44     52     67    145    119   ⋯\n",
       "6  426    357    515    597    631    588    590    604    881    576   ⋯\n",
       "  S2Q24C S2Q24D S2Q48A S2Q48B S2Q48C S2Q48D S2Q72A S2Q72B S2Q72C S2Q72D\n",
       "1   17     12     16     13     19     20      0     12     19     15  \n",
       "2 1986   1981   2208   2191   2095   2229   2013   2226   2125   2244  \n",
       "3    3     10     20      0     10     10      2      9      7     13  \n",
       "4   37      0     29     55     27     23     36     32     11     33  \n",
       "5  158    170    199    177    182    182    155    185    176    182  \n",
       "6  690    495    638    615    593    631    672    529    672    621  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "c <- read.csv(\"gene_count_matrix_sorted.csv\", row.names=1)\n",
    "gcts <- as.matrix(c)\n",
    "head(gcts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the same but for the transcript counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>R1C24A</th><th scope=col>R1C24B</th><th scope=col>R1C24C</th><th scope=col>R1C24D</th><th scope=col>R1C48A</th><th scope=col>R1C48B</th><th scope=col>R1C48C</th><th scope=col>R1C48D</th><th scope=col>R1C72A</th><th scope=col>R1C72B</th><th scope=col>⋯</th><th scope=col>S2Q24C</th><th scope=col>S2Q24D</th><th scope=col>S2Q48A</th><th scope=col>S2Q48B</th><th scope=col>S2Q48C</th><th scope=col>S2Q48D</th><th scope=col>S2Q72A</th><th scope=col>S2Q72B</th><th scope=col>S2Q72C</th><th scope=col>S2Q72D</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>MSTRG.9396.8</th><td>   0</td><td>   0</td><td>   0</td><td>   0</td><td> 16 </td><td>  0 </td><td>  0 </td><td>  0 </td><td>   5</td><td>  51</td><td>⋯   </td><td>  3 </td><td>  6 </td><td>  6 </td><td>  5 </td><td> 39 </td><td>  2 </td><td>  0 </td><td> 37 </td><td>  0 </td><td> 98 </td></tr>\n",
       "\t<tr><th scope=row>MSTRG.22509.1</th><td>   5</td><td>   2</td><td>   3</td><td>   2</td><td>  6 </td><td>  5 </td><td>  2 </td><td>  6 </td><td>   6</td><td>   7</td><td>⋯   </td><td> 10 </td><td> 22 </td><td> 19 </td><td> 19 </td><td> 14 </td><td> 13 </td><td> 22 </td><td> 20 </td><td> 10 </td><td> 21 </td></tr>\n",
       "\t<tr><th scope=row>MSTRG.2044.1</th><td>   0</td><td>   0</td><td>   0</td><td>   0</td><td>  0 </td><td>  0 </td><td>  0 </td><td>  0 </td><td>   0</td><td>   0</td><td>⋯   </td><td>  0 </td><td>  0 </td><td> 13 </td><td>  0 </td><td>  4 </td><td>  1 </td><td>  0 </td><td>  1 </td><td>  5 </td><td>  5 </td></tr>\n",
       "\t<tr><th scope=row>MSTRG.12668.1</th><td>1246</td><td>1183</td><td>1218</td><td>1168</td><td>824 </td><td>829 </td><td>816 </td><td>869 </td><td>1212</td><td>1287</td><td>⋯   </td><td>806 </td><td>908 </td><td>715 </td><td>657 </td><td>721 </td><td>669 </td><td>679 </td><td>689 </td><td>661 </td><td>654 </td></tr>\n",
       "\t<tr><th scope=row>MSTRG.18752.1</th><td> 297</td><td> 281</td><td> 313</td><td> 283</td><td>254 </td><td>225 </td><td>198 </td><td>182 </td><td> 295</td><td> 342</td><td>⋯   </td><td>398 </td><td>414 </td><td>359 </td><td>318 </td><td>313 </td><td>329 </td><td>317 </td><td>321 </td><td>358 </td><td>368 </td></tr>\n",
       "\t<tr><th scope=row>MSTRG.8651.1</th><td>   0</td><td>   0</td><td>   2</td><td>   2</td><td>  0 </td><td>  8 </td><td>  2 </td><td>  5 </td><td>   3</td><td>   6</td><td>⋯   </td><td>  7 </td><td>  3 </td><td>  0 </td><td>  0 </td><td>  0 </td><td>  3 </td><td>  4 </td><td>  2 </td><td>  3 </td><td>  3 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & R1C24A & R1C24B & R1C24C & R1C24D & R1C48A & R1C48B & R1C48C & R1C48D & R1C72A & R1C72B & ⋯ & S2Q24C & S2Q24D & S2Q48A & S2Q48B & S2Q48C & S2Q48D & S2Q72A & S2Q72B & S2Q72C & S2Q72D\\\\\n",
       "\\hline\n",
       "\tMSTRG.9396.8 &    0 &    0 &    0 &    0 &  16  &   0  &   0  &   0  &    5 &   51 & ⋯    &   3  &   6  &   6  &   5  &  39  &   2  &   0  &  37  &   0  &  98 \\\\\n",
       "\tMSTRG.22509.1 &    5 &    2 &    3 &    2 &   6  &   5  &   2  &   6  &    6 &    7 & ⋯    &  10  &  22  &  19  &  19  &  14  &  13  &  22  &  20  &  10  &  21 \\\\\n",
       "\tMSTRG.2044.1 &    0 &    0 &    0 &    0 &   0  &   0  &   0  &   0  &    0 &    0 & ⋯    &   0  &   0  &  13  &   0  &   4  &   1  &   0  &   1  &   5  &   5 \\\\\n",
       "\tMSTRG.12668.1 & 1246 & 1183 & 1218 & 1168 & 824  & 829  & 816  & 869  & 1212 & 1287 & ⋯    & 806  & 908  & 715  & 657  & 721  & 669  & 679  & 689  & 661  & 654 \\\\\n",
       "\tMSTRG.18752.1 &  297 &  281 &  313 &  283 & 254  & 225  & 198  & 182  &  295 &  342 & ⋯    & 398  & 414  & 359  & 318  & 313  & 329  & 317  & 321  & 358  & 368 \\\\\n",
       "\tMSTRG.8651.1 &    0 &    0 &    2 &    2 &   0  &   8  &   2  &   5  &    3 &    6 & ⋯    &   7  &   3  &   0  &   0  &   0  &   3  &   4  &   2  &   3  &   3 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | R1C24A | R1C24B | R1C24C | R1C24D | R1C48A | R1C48B | R1C48C | R1C48D | R1C72A | R1C72B | ⋯ | S2Q24C | S2Q24D | S2Q48A | S2Q48B | S2Q48C | S2Q48D | S2Q72A | S2Q72B | S2Q72C | S2Q72D | \n",
       "|---|---|---|---|---|---|\n",
       "| MSTRG.9396.8 |    0 |    0 |    0 |    0 |  16  |   0  |   0  |   0  |    5 |   51 | ⋯    |   3  |   6  |   6  |   5  |  39  |   2  |   0  |  37  |   0  |  98  | \n",
       "| MSTRG.22509.1 |    5 |    2 |    3 |    2 |   6  |   5  |   2  |   6  |    6 |    7 | ⋯    |  10  |  22  |  19  |  19  |  14  |  13  |  22  |  20  |  10  |  21  | \n",
       "| MSTRG.2044.1 |    0 |    0 |    0 |    0 |   0  |   0  |   0  |   0  |    0 |    0 | ⋯    |   0  |   0  |  13  |   0  |   4  |   1  |   0  |   1  |   5  |   5  | \n",
       "| MSTRG.12668.1 | 1246 | 1183 | 1218 | 1168 | 824  | 829  | 816  | 869  | 1212 | 1287 | ⋯    | 806  | 908  | 715  | 657  | 721  | 669  | 679  | 689  | 661  | 654  | \n",
       "| MSTRG.18752.1 |  297 |  281 |  313 |  283 | 254  | 225  | 198  | 182  |  295 |  342 | ⋯    | 398  | 414  | 359  | 318  | 313  | 329  | 317  | 321  | 358  | 368  | \n",
       "| MSTRG.8651.1 |    0 |    0 |    2 |    2 |   0  |   8  |   2  |   5  |    3 |    6 | ⋯    |   7  |   3  |   0  |   0  |   0  |   3  |   4  |   2  |   3  |   3  | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "              R1C24A R1C24B R1C24C R1C24D R1C48A R1C48B R1C48C R1C48D R1C72A\n",
       "MSTRG.9396.8     0      0      0      0    16      0      0      0       5  \n",
       "MSTRG.22509.1    5      2      3      2     6      5      2      6       6  \n",
       "MSTRG.2044.1     0      0      0      0     0      0      0      0       0  \n",
       "MSTRG.12668.1 1246   1183   1218   1168   824    829    816    869    1212  \n",
       "MSTRG.18752.1  297    281    313    283   254    225    198    182     295  \n",
       "MSTRG.8651.1     0      0      2      2     0      8      2      5       3  \n",
       "              R1C72B ⋯ S2Q24C S2Q24D S2Q48A S2Q48B S2Q48C S2Q48D S2Q72A S2Q72B\n",
       "MSTRG.9396.8    51   ⋯   3      6      6      5     39      2      0     37   \n",
       "MSTRG.22509.1    7   ⋯  10     22     19     19     14     13     22     20   \n",
       "MSTRG.2044.1     0   ⋯   0      0     13      0      4      1      0      1   \n",
       "MSTRG.12668.1 1287   ⋯ 806    908    715    657    721    669    679    689   \n",
       "MSTRG.18752.1  342   ⋯ 398    414    359    318    313    329    317    321   \n",
       "MSTRG.8651.1     6   ⋯   7      3      0      0      0      3      4      2   \n",
       "              S2Q72C S2Q72D\n",
       "MSTRG.9396.8    0     98   \n",
       "MSTRG.22509.1  10     21   \n",
       "MSTRG.2044.1    5      5   \n",
       "MSTRG.12668.1 661    654   \n",
       "MSTRG.18752.1 358    368   \n",
       "MSTRG.8651.1    3      3   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "c <- read.csv(\"transcript_count_matrix.csv\", row.names=1)\n",
    "tcts <- as.matrix(c)\n",
    "head(tcts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, I create the sample the sample information table as a data frame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>exp</th><th scope=col>cond</th><th scope=col>hpi</th><th scope=col>rep_bio</th><th scope=col>rep_tec</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>R1C24A</th><td>R  </td><td>C  </td><td>24h</td><td>1  </td><td>1  </td></tr>\n",
       "\t<tr><th scope=row>R1C24B</th><td>R  </td><td>C  </td><td>24h</td><td>1  </td><td>1  </td></tr>\n",
       "\t<tr><th scope=row>R1C24C</th><td>R  </td><td>C  </td><td>24h</td><td>1  </td><td>1  </td></tr>\n",
       "\t<tr><th scope=row>R1C24D</th><td>R  </td><td>C  </td><td>24h</td><td>1  </td><td>1  </td></tr>\n",
       "\t<tr><th scope=row>R1C48A</th><td>R  </td><td>C  </td><td>48h</td><td>1  </td><td>2  </td></tr>\n",
       "\t<tr><th scope=row>R1C48B</th><td>R  </td><td>C  </td><td>48h</td><td>1  </td><td>2  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllll}\n",
       "  & exp & cond & hpi & rep\\_bio & rep\\_tec\\\\\n",
       "\\hline\n",
       "\tR1C24A & R   & C   & 24h & 1   & 1  \\\\\n",
       "\tR1C24B & R   & C   & 24h & 1   & 1  \\\\\n",
       "\tR1C24C & R   & C   & 24h & 1   & 1  \\\\\n",
       "\tR1C24D & R   & C   & 24h & 1   & 1  \\\\\n",
       "\tR1C48A & R   & C   & 48h & 1   & 2  \\\\\n",
       "\tR1C48B & R   & C   & 48h & 1   & 2  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | exp | cond | hpi | rep_bio | rep_tec | \n",
       "|---|---|---|---|---|---|\n",
       "| R1C24A | R   | C   | 24h | 1   | 1   | \n",
       "| R1C24B | R   | C   | 24h | 1   | 1   | \n",
       "| R1C24C | R   | C   | 24h | 1   | 1   | \n",
       "| R1C24D | R   | C   | 24h | 1   | 1   | \n",
       "| R1C48A | R   | C   | 48h | 1   | 2   | \n",
       "| R1C48B | R   | C   | 48h | 1   | 2   | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "       exp cond hpi rep_bio rep_tec\n",
       "R1C24A R   C    24h 1       1      \n",
       "R1C24B R   C    24h 1       1      \n",
       "R1C24C R   C    24h 1       1      \n",
       "R1C24D R   C    24h 1       1      \n",
       "R1C48A R   C    48h 1       2      \n",
       "R1C48B R   C    48h 1       2      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colcols <- c(\"exp\", \"cond\", \"hpi\", \"rep_bio\", \"rep_tec\")\n",
    "\n",
    "coldata <- read.csv(\"pheno_data.csv\", row.names=1, strip.white=T)\n",
    "coldata <- coldata[, colcols]\n",
    "\n",
    "# Correct the technical replicate column for later collapsing\n",
    "coldata$rep_tec <- c(rep(1:4, each=4), c(5,5,5), rep(6:24, each=4))\n",
    "head(coldata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just checking for consistency between the rows of the sample information table and the columns in the count matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check with the gene counts\n",
    "all(rownames(coldata) %in% colnames(gcts))\n",
    "# Check with the transcript counts\n",
    "all(rownames(coldata) %in% colnames(tcts))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, the DESeqDataSet objects can be created. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      Length        Class         Mode \n",
       "       24413 DESeqDataSet           S4 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "      Length        Class         Mode \n",
       "       55493 DESeqDataSet           S4 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gdds <- DESeqDataSetFromMatrix(countData = gcts,\n",
    "                               colData = coldata,\n",
    "                               design = ~ cond)\n",
    "tdds <- DESeqDataSetFromMatrix(countData = tcts,\n",
    "                               colData = coldata,\n",
    "                               design = ~ cond)\n",
    "summary(gdds)\n",
    "summary(tdds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I'll also load an annotations file containing the information of chromosome and position for each gene id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>gene_id</th><th scope=col>chr</th><th scope=col>g_star</th><th scope=col>g_end</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>1     </td><td>0     </td><td> 29755</td><td> 30541</td></tr>\n",
       "\t<tr><td>2     </td><td>0     </td><td> 68426</td><td> 86044</td></tr>\n",
       "\t<tr><td>3     </td><td>0     </td><td> 97938</td><td> 99413</td></tr>\n",
       "\t<tr><td>4     </td><td>0     </td><td> 98077</td><td>101005</td></tr>\n",
       "\t<tr><td>5     </td><td>0     </td><td>182612</td><td>193163</td></tr>\n",
       "\t<tr><td>6     </td><td>0     </td><td>196283</td><td>206073</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       " gene\\_id & chr & g\\_star & g\\_end\\\\\n",
       "\\hline\n",
       "\t 1      & 0      &  29755 &  30541\\\\\n",
       "\t 2      & 0      &  68426 &  86044\\\\\n",
       "\t 3      & 0      &  97938 &  99413\\\\\n",
       "\t 4      & 0      &  98077 & 101005\\\\\n",
       "\t 5      & 0      & 182612 & 193163\\\\\n",
       "\t 6      & 0      & 196283 & 206073\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "gene_id | chr | g_star | g_end | \n",
       "|---|---|---|---|---|---|\n",
       "| 1      | 0      |  29755 |  30541 | \n",
       "| 2      | 0      |  68426 |  86044 | \n",
       "| 3      | 0      |  97938 |  99413 | \n",
       "| 4      | 0      |  98077 | 101005 | \n",
       "| 5      | 0      | 182612 | 193163 | \n",
       "| 6      | 0      | 196283 | 206073 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  gene_id chr g_star g_end \n",
       "1 1       0    29755  30541\n",
       "2 2       0    68426  86044\n",
       "3 3       0    97938  99413\n",
       "4 4       0    98077 101005\n",
       "5 5       0   182612 193163\n",
       "6 6       0   196283 206073"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "annot <- read.csv(\"annotations.txt\", header=F)\n",
    "colnames(annot) <- c(\"gene_id\", \"chr\", \"g_star\", \"g_end\")\n",
    "# Re-order annotation gene_ids according to the order in the data set\n",
    "annot <- annot[match(rownames(gdds), annot$gene_id),]\n",
    "head(annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collapse technical replicates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DESeq2 provides a function collapseReplicates which can assist in combining the counts from technical replicates into single columns of the count matrix. The term technical replicate implies multiple sequencing runs of the same library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gdds <- collapseReplicates(gdds, gdds$rep_tec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mcols(gdds) <- cbind(mcols(gdds), annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data exploration\n",
    "\n",
    "DESeq2 offers some ways of exploring the data before undertaking the differential gene expression analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdd3xUVf7/8XunZUoyKYQWIBCqkRJCqAoCIggiCmJlf1Rpuusu6IMiLq4F\nBVERhQVBQGw0Gy5YEJAisO7SpQgYIHQhIWUyTDL1/v6YNd+YQOYmZO4kd17PP3xkzpzc+5m9\nbPLOufecI0qSJAAAAKD604S6AAAAAFQOgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAA\nqATBDgAAQCUIdgAAACpBsAMAAFAJgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATB\nDgAAQCUIdgAAACpBsAMAAFAJgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAA\nQCUIdgAAACpBsAMAAFAJgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUI\ndgAAACpBsAMAAFAJgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAA\nACpBsAMAAFAJgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpB\nsAMAAFAJgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMA\nAFAJgh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFAJ\ngh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFAJgh0A\nAIBK6EJdQAU5HA6HwxHqKhQSHR1tt9u9Xm+oC1GCwWCwWq1hdX3j4uKys7NDXYVCjEZjZGRk\nfn6+0+kMdS1KEEUxJiYmJycn1IUoxGKxmEym3Nxcj8cT6lqUoNPpzGazzWYLdSEKsVqtBoMh\nOzvb5/NV7pHj4+Mr94DhjBE7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFAJgh0AAIBK\nEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFAJgh0AAIBKEOwA\nAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFAJXagLQMiITqfouCYZ\nDJIlMtS1AACASkCwCz+SpD9+VL/739rLlwRJEgRBio5xt2rr6tBF0utDXRwAAKg4gl2Y8XqN\n36zVHztSvE3MyzXs3Ko7drjgwT/5rNGhKg0AANwknrELL8YfNpRIdUU0V7NMn68QPW6FSwIA\nAJWFYBdGtJcv6Q/sKaODJivTsOc/itUDAAAqF8EujOgPHwzYR/fzPgUqAQAAwUCwCyPa3y4E\n7KPJyxULChQoBgAAVDqCXThxOuX0El2yugEAgKqGYBdGJEtU4E4ajWSxBL8WAABQ+Qh2YcTT\nMClgH29CfUnHanYAAFRLBLsw4k5JkyKMZfdxdeiiTDEAAKDSEezCiGQyOe++t4wO7tapnqYt\nFKsHAABULoJdeHG3uLXw/ockY6lxO63W1aFLYZ/+oSgKAABUDrYUCzvu5smeBo30h/Zrz5wW\n7flCRIQ3ob67VYovvlaoSwMAADeFYBeOJJPJ1fE2oeNtoS4EAABUJm7FAgAAqATBDgAAQCUI\ndgAAACpBsAMAAFAJgh0AAIBKEOwAAEBZ8rzeE07XJZdbCnUlCIjlTgAAwHVIgvBVXv6CrOwD\nBYX+SFc/wvBotPXP8bGRGgaGqiguDAAAKMktSePPXRpz7uL+31OdIAjnna43rmTddfJMhssd\nyuJwYwQ7AABQ0nOXrnyRZ7vuWyedrkfPnLf7fAqXBDkIdgAA4A8OFTrfz84to8NJp+ufWdmK\n1QP5CHYAAOAPVuXkBe6Ta2MuRRVEsAMAAH+wv6AgYJ/zLnemx6NAMSgXgh0AAPgDu1fW83P5\n8rpBSQQ7AADwB3X0gVdD0whCbRndoDCCHQAA+IPukZaAfdLMRlazq4K4JAAA4A/+FBsdo9WW\n3efP8XHKFINyIdipgZifrzv1q/7EL5qL5wUWFgIA3JwYrfbterXL6PBojLW/NUqxeiAfd8er\nN23m5YitG7VnTgvS/2adSyaTq+NtrvZdBEbIAQAVdY816sPEehMu/pbt8RZv14vimBqx02vH\nh6owlI1gV43p0o8b130u/nG2uVhQELFts+5MRsEDj0hari8AoIL6WSNvsyStzLVts1/7ze2J\nMRg6R1sHmSKaG/ShLg03xC/+6kqTlWlc94V4gzWEtBknIzZ+W9h3gMJVAQDUJFqrHV8jdnyN\nWEEQrFarwWDIzs728cxPFcbduuoqYucWwVPWHsz6Q/u1mVcUqwcAAIQcwa5aEp1O3clfxUDd\ndMcOK1ENAACoGgh21ZImN1vwegN3Y8QOAIBwQrCrnmSkOkEQRJ+sbgAAQB0IdtWSzxotr1tM\nsCsBAABVB8GuWpIio7y16wbs5mnSTIFiAABAFUGwq65cXbqV3cFbJ8HTuNzBTnS7RXs+21cA\nAFAdsY5ddeVpdosrrZNh73+u+65kMhcOGCyIASfO/o/odhv2/kd3+KAm56ogCJJW601s5Op4\nuzexUWUVDAAAgo1gV40577xbskYbdmwV3a7i7d76iYX3DPRFy33ATpOXa/pshSY7q6hF9Hp1\np0/qTp90deji7H6X/IAIAABCiGBXvbnad3a3TNH9+ov28m+iy+W1Wr2Nm3nrNZB/BNFZaPr0\nE/9AXWmG3f8WjCZn566VVC8AAAgigl21J5lM7jbtytqDokyG/+66Uarz0+/a5k5uJX/8DwAA\nhAqTJ8KbJOl/3l92F9Hr1R85qEw5AADgZhDswprGlic6rgXudvG8AsUAAICbRLALa2JhgZxu\nGqcz2JUAAICbR7ALaz5LpKxukbK6AQCA0CLYhTUpMsoXVyNgN0+DRsGvBQAA3CyCXbhzte9c\ndgfJZPK0bKNMMQAA4GYQ7MKdu3WqJ6lpGR0K+94nRRgVqwcAAFSYKElSqGuoCJfLpdGESyrV\narU+ny+IV8rt8n35qXRwryAIgiQIRdtMGE2aBx4RlR2uE0XR/3l9YbNfrU6n83g8oa5CIRqN\nRqPReL3eavqTpwK0Wq3X6w11FQoJt+sriqL/84a6EIVotVpRFIPx80qnY1XdSlNdg53D4Sgo\nkDWjUwWsVqvdbg920NGeP6M/fFBz6aLgckpRVm+jJq627QWzOagnLU2v11ut1oKCAofDofCp\nQyU2NjYnJyfUVSjEaDRaLBa73e4Mj6nWoihGR0fn5uaGuhCFWCwWo9GYl5cXJn+raLVas9mc\nn58f6kIUEhUVZTAYcnJyKv33UY0agZ/2hkzVOCNX00haYcH+vJ56iZ56iaXPGtST3ogkSWF1\nfcPnw/o/KddXrbi+4SDcrm+1Ey53MwEAAFSPYAcAAKASBDsAAACVqMbP2KEsPp/++FHtr8c0\nOVcFjVaKi3e3SPY0aS6IYuDvBQAA1RPBToU0OdnGtWu0WVcE4fflS367qDv6s7deg4L7HpQi\no0JdIAAACApuxaqNmJdrXvH+/1KdUGxROkHQXjhnXvWBGDbLxAAAEG4Idmpj/P5r0XHtRu9q\ncrIjtm1Ush4AAKAYgp2qaDKv6DJOlt1Hf+RnsSBc1v4FACCsEOxURXf2dOBOPp/ubEbQSwEA\nAIoj2KmKaLfLWQ5ctIfLBjgAAIQVgp26RETIWc5EiogIeiUAAEBxBDtV8dSpK6ebt05CsCsB\nAADKI9ipirdBI190TNl9fAn1fPG1lKkHAAAoiWCnLlqt8657ynhf0ukKe/VTrBwAAKAkgp3a\neBo3Lew7QNJqS78l6Q2FAwZzHxYAALViSzEVcrdO9dWtp9+5TXcqXfS4BUGQIiK8TW9x3t49\n4I1aAABQfRHs1MkbX8t7/0Oi1yva8yVBkCKjhOuN4QEAADUh2KmZpNVKDNEBABA2eMYOAABA\nJQh2AAAAKkGwAwAAUAmCHQAAgEoQ7AAAAFSCYAcAAKASBDsAAACVINgBAACoBMEOAABAJQh2\nAAAAKkGwAwAAUAn2igUAQLUkQThS6MxwubWC0Nigb2GMCHVFCC6CHQAA6vRZrm3m5ayzbndR\nyy3GiOm14/tERYawKgQVt2IBAFChZy9deeL8peKpThCEY4XOP525MCfzaqiqQrAR7AAAUJsl\nV3OWXM250bszL2ett9mVrAeKIdgBAKAq13y+164EGJN7/tJlryQpUw+URLADAEBVttgduV5v\n2X3OuT17CwqVqQdKItgBAKAqxwqdcrodldcN1QvBDgAAVXFKPjndXNyKVSOCHQAAqtLQYAjQ\nQxIEQWhk0CtQDBRGsAMAQFXujLToRLGsHqJg1mi6WsxKVQTlEOwAAFCVBL1uSGx02X3+HB9r\n1pABVIiLCgCA2rxUp2Ybk/FG73aPtEysWUPJeqAYgh0AAGpj0Wi+SmrwcEzJcTu9KI6Oi1nR\nsJ6+7Hu1qLbYKxYAABWK1Gj+Wb/OhJpx39ns6S6XVhCaGyP6W6Ma6PnVr2ZcXQAAVKtZhKFZ\nzbhQVwHlcCsWAABAJQh2AAAAKkGwAwAAUAmCHQAAgEoQ7AAAAFSCYAcAAKASBDsAAACVINgB\nAACoBAsUhwvNhXO6U79qbHmCRuOLr+VunixFx4S6KAAAUJkIduonXrObvvlKm3GyeGPEjz+4\n2nV03tFL0DBqCwCAShDsVE68Zrd8skzMyy35htdr2P1vTV5OwX0PCWwFDQCAKjBao3LGb/91\nnVT3O92JY4b9e5SsBwAABA/BTs20v13UnU4vq4ckGP69XfD5lKoIAAAEEcFOzXSnykx1giCI\ngui4pvntkiLlAACA4CLYqZlou+FN2OK08roBAIAqjmCnajq9nF6Sjjk0AACoAcFOzbzxNeV0\n89WsHexKAACAAgh2auZpdkvA0ThfvQY+VioGAEAVCHZqJlki3Z26ltVBqy3ofpdi9QAAgKAi\n2Kmcs0s3T6uU674labWFfe711WugcEkAACBIeGpe7USxoN/9+gYNDTu3aWx5RY3eeg0Ku/f2\nJdQLaXEAAKAyEezCgrtVW3fLFO3VTDE3R9BqvfG1paioUBcFAAAqGcEubIiiN76WEF8r1HUA\nAIBg4Rk7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFCJEMyKdblcy5Yt279/f15eXrNm\nzUaNGpWUlCQIwhdffLF8+fKiblqt9ssvv1S+PAAAgGoqBMHutddeO3369Lhx42JiYlatWvXC\nCy/885//jIyMvHz5crt27e677z5/N1EUla8NAACg+lI62GVlZe3evfv5559v3769IAhTpkwZ\nOnTonj17evTocfny5VtuuaVdu3YKlxRGJEmbdUVzNUuQJF9cDW+tOgLpGQAAFVE62NlstqZN\nm7Zo0cL/MiIiwmg05ubmCoJw+fLlHj16FBYWut3uKPZFqGz6E78Ytm3W5GYXtfis0c5uPT23\ntglhVQAAoBIpHewaN248Z86cope7d+/Oy8tr2bKlJEmXL19ev379W2+9JUlSgwYNnnrqqVtu\nuaWop8vl2r59e9HL+vXr16sXLvucajQag8Hg8/kqfATt1o3aHVtLHtaWZ/p6rffKb967B9xU\nfZVKp9P5/xsRERHqWhQiimL4fFj/9dXr9aEuRDlhdX21Wq0gCAaDwf+F6mk0Go1GEz7XV6PR\nCIJgMBgkSQp1LbghMVSXR5KkjRs3Llq06O677x47duzVq1fHjRvXr1+/wYMHezye999//8CB\nAwsWLIiOjvb3z8nJ6d27d9G3jx07duzYsSGpvNrx7t/tWfVRGR10gx7Wdu6qWD0AgGDzScKW\n3NwtuXlXXO4onbaL1dq/RqxJw1IY6heaYHf58uW33norIyNj+PDh/fr1K93B6XQOHTp0/Pjx\nd955p7+lsLBw9erVRR1atmyZnJysULmhZjQaXS5XBUfsvF7dvNmCzVZWH5PZM2GKoDdUrLzK\npdVqjUaj2+12uVyhrkUhZrPZ4XCEugqF6PV6g8HgdDo9Hk+oa1GCKIpGo7GgoCDUhSjEYDDo\n9fqCgoKbucNQjfhvpxQWFoa6kJJ+dhSMO332kOMP//DqGvRvJNa/Pza6woc1Go1ardbhcFR6\ncrBYLJV7wHAWglmxJ06cmD59eps2bRYtWlQ0IFdCREREzZo1/c/e+RmNxuHDhxe9dDgc4fO7\n0P+Dw+v1VuB7tWdP68pOdYIgFDjcx456mraoSHGVzWAw+INd+PwuNJlM4fNhJUkyGAwul8vp\ndIa6FiX478OGz/XVaDR6vT58grtOp9NqtVXt+u5xFAzOOO8ola0vudx/Sj/9RkLt4XExFTuy\nXq/XarWFhYWVHtwJdpVI6VFZr9c7c+bMHj16TJs2rXiq27t375///Gfb7xHE4XBcuXIlMTFR\n4fLUR5OVKQiB/7TSZmUqUAwAIKgKfNKocxdLp7oi0y5d+aUwLP6sCltKj9jt378/Ozs7JSXl\nyJEjRY0JCQmtW7e22+1z5swZOHCgXq9ftWpVgwYNWPrk5ok+nyAEXtNEqtBwIACgSvk4J++S\nu6zhUpckvZV5dXGDBMVKgsKUDnbnz5+XJGnWrFnFG8eNG9e/f/833nhjyZIlb775plarbdeu\n3eTJkzU85nnTfNGyhtx9MbHBrgQAEGzf59sD9tmYf80nCRqWMVUppYPdwIEDBw4ceN23atas\n+eyzzypcj+p5E5MknV70uMvqpNV6k5ooVREAIFjOyJh2Zvf5rnq9NXVhsSRNGGJITOWkiAhX\nh85l93G1bS+ZeXAVAKo9g7w7XXqG69SLYKd+ri53eBsm3ehdb0J9V/deStYDAAiS5obAq3/X\n1etiwmMF6fBEsAsDWm3B4CGudh2FEv9P1mrdKWkFjwyVtCFY9QYAUOkeiAm8TN2gaKsClSBU\n+I0eFiSt1tmrr6tDF336cU32VUGSfLFxnma3yJxaAQCoFvpbIzuaTf913HBpvZo67V9rxilZ\nEhRGsAsjkjXa1a5jqKsAAASLKAjvJyY8kHH+uH+xOukPC17F6rQfJNarwX1YVeNWLAAA6lFL\np/s2KXFMjZgIUSxKdRpBuMcatbFxww5mU0irQ9AxYgcAgKpEaTWv1q3999o1/+souOz2xOi0\n7Uwm1jcJEwQ7AABUyKzR9IhkKauww61YAAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFAJ\ngh0AAIBKEOwAAABUgmAHAACgEgQ7AAAAlSDYAQAAqATBDgAAQCUIdgAAACpBsAMAAFAJgh0A\nAIBKEOwAAABUgmAHAACgErpQFwAAACqZW5LW5Nq+zMs/7nR6JCnRoO8bFTUqLjpaqw11aQgu\ngh0AAKpy1u0eeubC0UJnUUuWx7vPUfhuVvaSBgndIs0hrA3Bxq1YAADUI9vrHXTqXPFUV/yt\nIWfO73MUKl8VFEOwAwBAPV76LfOs232jdwslacLF37ySpGRJUBLBDgAAlbB5fZ/m2cru80uh\nc7vdoUw9UB7BDgAAldhdUODyBR6N2+Ug2KkWwQ4AAJXIcnvkdMv0eINdCUKFYAcAgErE6mSt\nZhInrxuqI4IdAAAqkWY2aUUxYLeOJpMCxSAkCHYAAKhEDa22f1Rk2X0S9fo7oyzK1APlEewA\nAFCPF+vUrFHm9hJvJNQ2yBjVQzVFsAMAQD3qG/SfJjWop9eXfssoiosbJPRkuE7VCHYAAKhK\na2PEj00bTa4V3zzC4B+aq6PXjYiL2dW88aDoqBAXhyBjr1gAANQmSquZVKvGpFo13JLkFQQj\n917DBsEOAADV0ovidW7KQr24FQsAAKASBDsAAACVINgBAACoBMEOAABAJQh2AAAAKkGwAwAA\nUAmCHQAAgEqwjh0AAOqR7fFe9nqtGrGuTq9hWeLwQ7ADAKDakwThyzzbP7Nyfi4o9LfU0uke\njY3+a3xstFYb2tqgJG7FAgBQvbklafy5S+POXfpfqpMEQRCueDzvZF7tdfJMutMV2vKgJIId\nAADV2/OXrnyRZ/u/18XuwJ5xuR87c97m9SlfFUKCYAcAQDV2tNC5JDu3jA4ZLve8rGzF6kFo\nEewAAKjGVuXmBeyzMidPUqAUVAEEOwAAqrEDv8+WKMNlj+eS26NAMQg5gh0AANWYXd7zc/le\nb7ArQVVAsAMAoBqrq9cH7CMKQh09C5yFBYIdAADVWI9IiyD8b4mTG2lnNrGaXZgg2AEAUI09\nGmOtodUKZW4y8Zf4OKXKQYgR7AAAqMaitJp/Nqj7vxfXG7d7NDb6XmukkiUhhAh2AABUb70i\nLasa1S89bqcXxSfj495KqB2iuhACPEoJAEC11yvSsrt549V5tm32a7+5PRaNpp3Z+HC09RZj\nRKhLg6IIdgAAqEGUVjM6LmZ0XEyoC0EocSsWAABAJQh2AAAAKkGwAwAAUAmCHQAAgEoQ7AAA\nAFSCYAcAAKASBDsAAACVINgBAACoBMEOAABAJQh2AAAAKkGwAwAAUAmCHQAAgEqIkiSFuoaK\ncLlcWq021FUoRKPR+Hy+UFehEFEUNRqNJEnh85G1Wq3X6w11FQrxX1+fz1dNf/JUQFhdX41G\nI4pi+FxfURT9nzfUhSjEf32D8e85fH6hK0AX6gIqyOPx2Gy2UFehkOjoaLvdHia/GwwGg9Vq\nLSgocDgcoa5FIXFxcTk5OaGuQiFGozEyMvLatWtOpzPUtShBFMWYmJjwub4Wi8VkMtlsNo/H\nE+palKDT6cxmc/j8MrJarQaDIS8vr9KzbHx8fOUeMJxxKxYAAEAlCHYAAAAqQbADAABQCYId\nAACAShDsAAAAVIJgBwAAoBIEOwAAAJUg2AEAAKgEwQ4AAEAlCHYAAAAqUb4txfLz83/66aes\nrKyePXtGRUWZzWZRFINUGQAAAMqlHCN2ixYtqlu3bp8+fYYMGXL8+PG1a9c2bNjw008/DV5x\nAAAAkE9usFu/fv348eM7dOiwcuVKf0v79u0TEhIeeeSRDRs2BK08AAAAyCU32L322mtt27bd\ntGnTo48+6m9p0aLF9u3bU1NTZ86cGbTyAAAAIJfcYHfw4MFBgwZptdrijQaD4ZFHHjl48GAQ\nCgMAAED5yA12sbGxhYWFpdsvXrwYFRVVqSUBAACgIuQGu86dO3/00Ue5ubnFG0+dOrVq1aqO\nHTsGoTAAAACUj9zlTl577bWUlJTU1NQxY8YIgrBp06YtW7YsWrTI4XDMmjUrmBUCAABAFrkj\ndo0aNdqxY0ejRo2ee+45QRBmzJjx4osvtm7d+scff2zatGkwKwQAAIAs5ViguHXr1lu2bMnJ\nyTl+/LjBYGjSpEl0dHTwKgMAAEC5lCPY5eTkfPjhh23btu3evbsgCO+++67dbh8zZgzxDgAA\noCqQeyv26tWr7dq1mzBhwtGjR/0tx44dmzRpUkpKyrlz54JWHgAAQHCVmBt6MzwejyRJlXW0\nCpAb7CZNmpSTk7Np06YnnnjC3zJ37twdO3bYbLZp06YFrTwAAIAK+vHHH3v16hUTE9OlS5dP\nP/10zJgxqamp/rd69uz50EMP/frrr3379m3fvr2/cd++ff37969Tp05CQkL//v337dtXdKjU\n1NQBAwYUP/iAAQNat24tCILX6xVFcd68ecOGDTObzWaz+bbbbvvwww+V+pR/IPdW7Pbt28eM\nGdOrV6/ijbfffvu4ceM++eSTIBQGAABQcVu2bOnXr19ycvIzzzyTkZExZMiQ+Pj4OnXqFHXI\ny8sbNGhQWlragw8+KAjC5s2b77nnnjp16owcOVIQhBUrVnTp0uXrr7++66675Jzu5Zdfttls\nI0aMqFWr1tq1a4cPH37x4sWpU6cG6dPdiNxgl5OTExkZWbrdYrHY7fZKLQkAAOBmTZ48uUmT\nJrt27TKZTIIgdOrUady4ccWD3caNGxcvXuxfx83n8z399NPx8fF79+6Nj48XBOHpp59OSUmZ\nNGnSvn37RFEMeLrMzMzNmzffeeedgiA899xzvXr1euWVVx5//PGaNWsG6xNej9xbsWlpaZ99\n9pnD4SjeWFhY+Nlnn7Vr1y4IhQEAAFTQqVOn9uzZM3bsWH+qEwRh5MiRVqu1eB+LxTJq1Cj/\n16dPn/7555+ffPJJf6oTBKFmzZrjx48/cODAmTNn5Jzxjjvu8Kc6QRBMJtP06dPtdvv3339f\nOZ9HNrkjdi+++GL37t27dOny17/+tWXLljqd7tixY3Pnzj106NCmTZuCWiIAAEC5pKenC4LQ\nvHnzoha9Xp+UlFS8T2Jiolar9X998uRJQRBatWpVvIP/EbqTJ082atQo4BnbtGlT/KX/YT7/\nYZUkN9h16dLlyy+/nDhx4ujRo4sa69Wr98knn/Ts2TM4tQEAAFSE0+kUBKHELVStVuvz+Ype\nWiyWEt9Vor9GoxEEwePxXPcUXq+3jO/V6XSCILhcrvJWfpPKsY5d//7977777n379qWnp7tc\nrqZNm6alpRWNcAIAAPnckvRzQeFljydWq2tljIjSyn04CnL4t8U6ceJE3759/S0ej+f06dMN\nGza8bv8mTZoIgnD48OH77ruvqPHw4cNCsWG/EuuYZGRkFA34CYJw6NCh4u8ePHhQ+OOQoTLK\n989Ip9N17NhxyJAhI0aM6Nq1K6kOAIDyKvBJMy9nJR872ffU2eFnL953+uwtx9OfOH/pgtsd\n6tLUo3nz5snJye+9915hYaG/5aOPPsrJyblR/6SkpFatWi1YsCA7O9vfcvXq1YULF7Zq1cqf\nBU0m07Fjx4pG6bZu3frLL78UP8LWrVu3bdvm/9rpdL700ktGo7HEciIKkDtiZ7PZnn766Y0b\nN5aYP+GXmZlZqVUBAKBOOR7vw2fOHygoLN7o8kmf5dq25F9b3ah+iskYqtrURKvVzps3r2/f\nvt26dRs8ePCZM2fWr1/fpEmT0rdf/TQazZw5c/r379++ffuhQ4dKkvTRRx9lZv+SBS0AACAA\nSURBVGZ+8MEH/huyd9555yuvvPLAAw888MADJ0+enDNnTolD1atXr1+/fqNGjapZs+aXX355\n8ODBl156qV69ekp82mLkBrtnnnlm6dKl7du3b9Omjf8TAgCA8hp34VKJVFfkqtf7/85e2N6k\nUaxOe90OKJdevXpt2rRp2rRps2bN6tChw4YNGx577LHiy52U0Lt37127dk2fPn3RokWCILRr\n1+6LL74oWtB4+vTpDodjzZo169atkyTpT3/6U3x8/ObNm4u+feTIkY0bN54/f/6vv/6anJy8\nbNky/3p4CpMb7NatW/fwww+vXr06qNUAAKBim/Kvbcm/dsO3JeE3t+edrOx/1FF05TNVkiTp\nvffea9Wq1c6dO/0t+fn5p06d6tOnj//lli1bSn9X+/btv/322+seMCIiYs6cOXPmzLl27VpB\nQUHRqijFjRw5MiRhrjhZY29ut/vy5ctF/1sAAIAK+DzPVtbb4v/6hHKrUbUQRXHFihWPPfbY\n5s2b8/PzMzIyxo0b53a7n3rqqZs8ssViuW6qqyJkjdhptdqaNWv653cAAICKOVroDNjnktuT\n4/HGcTf2pn388cePPvpo0YZg9erVW7duXWJiYmirCjZZwU6j0bzzzjsjR45s06bNqFGjeMYO\nAIAKcEqyBuNkdkPZ6tevv2PHjpMnT545c6Zhw4ZJSUlBCjCiKP7lL3/p1KlTMA5eXnKfsVu9\nenXdunXHjBkzceLEhg0b6vX64u/u378/CLUBAKAqDfT6k84AK9YaRbEmw3WVp0mTJv416oJH\no9HMmzcvqKeQT26wKywsbNGiRYsWLYJaDQAAKtY7yrLVfuPJE4IgCMKdUZE6GVvOA9clN9jd\naJIIAACQ6U+x0e9kZl++wRZVfhNrxilWD9TnZm82b968uWizDgAAUAaLRvNegwTjjQfk/lGn\nZlsWKMZNKMdesWvWrNm0aVNBQUHxxl27dtnt9squCgAAdepiMa1NavCXC7+l//Fhu1id9qU6\ntR6NsYaqMKiD3GC3ePHicePGWa1Wj8fjcDgaNmzo9XovXLhQu3btN998M6glAgCgJmlm049N\nG23Kv/bjNcdlj8eq0aSZTfdaI6O1zJnAzZIb7BYsWJCSkvLf//43MzMzMTFx27ZtDRs23LFj\nx6BBg3r06BHMCgEAUBudKPa1Rva1Roa6EKiN3GfsTp482bdvX4PBUK9evbZt2+7Zs0cQhK5d\nuz7wwANTpkwJZoUAAADlJkmSr1KF+gPJInfEzmQyFX2k1NTUHTt2DB48WBCETp06vfjii8Gq\nDgAAoEIKCgq8Xm8lHtBsNmur/O1yuSN2ycnJGzZscLlcgiCkpKR89dVX/vbjx4/bbGXufAcA\nAABFyA12U6ZM+fnnn5s1a3bt2rWuXbtmZGSMHz9+7ty5S5cu7dKlS1BLBAAAgBxyb8Xec889\ny5cvX7FihSRJqampL7744ssvv+x2u5OSkubMmRPUEgEAACCHKFV0p+H8/PyzZ882b968xL6x\nynA4HA6HQ/nzhkR0dLTdbq/cBwWqLIPBYLVaw+r6xsXFZWdnh7oKhRiNxsjIyPz8fKfTGepa\nlCCKYkxMTE5OTqgLUYjFYjGZTLm5uZ4yd1ZQDZ1OZzabw+d5JKvVajAYsrOzK30aQXx8fOUe\n0M/hcIThM3blWKC4hKioqJYtW1ZiKQAAALgZZQW73r17WyyWtWvX+r8uo+fGjRsruS4AAIBg\n83qFa3bRYJCMplCXUjnKCnbFh9PZNwwAAKiGeOaUdtd2TcYpwesVBEGKjvG1aefr3FWKiAh1\naTelrGC3e/fuoq///e9/B78YAACAIJMk7Zbvtbu2/6ExL1f74w+aQ/s9jw6X4mvKPNLQoUM/\n/vjjopdGo7FFixbTpk17+OGHy1VRYWHhSy+99N133x0/frx27dp33333Cy+8ULt27XIdxE/W\ncie5ublNmzZdunRpBU4AAABQdWh/2lEy1QmC6P9vbo5u5XKhoByz9zp37vzT79atW9esWbPH\nHnts79698o9QUFDQsWPH1atXjx49+l//+te0adO2bt3au3fvik0ilDV5IiYmpkmTJjt27Hj8\n8ccrcA4AAIAqwWbTbt9cxvtiXq52+w/eu++VebyYmJhOnToVvbzjjjuio6M3btyYlpYm8wiv\nvfaazWbbt29fXFycv+X+++9v3LjxihUrRo8eLfMgReQuUDx//vzdu3cvWbIkTBbdAAAA6qM9\nckBwu8vuozm0X6ho2jEYDEajsUaNGqXfWrp0qSiKJ0+eLN4oSdKbb745adKkolQnCELNmjXf\nf//9xMTEChQgd7mTadOmJSQkjBkzZuLEiYmJiWazufi7xZ/GAwAAqJrEixcC9yksFK9mSrXq\nlPfgNpvt3XffNRqN/fr1K/1umzZtpk6dGhMTU7wxIyPDbrffdtttJTo/+OCD5T27n9xgl5WV\nJQhCjx49KnYaAACAkBMLC2T1KyyUecDvvvtOFMWil1qtduvWrfXr1y/ds0OHDh06dCjRmJGR\nIQhC3bp1ZZ4uILnBbsuWLZV1SgAAgJCQLFFi4F6CEGWVecDOnTvPnTvX/7Xdbp83b97gwYPP\nnTtnMBjkfLv/fuulS5fq1PnDAGFWVpZ/6qrMMopUfOcJv82bN7/++uvffffdTR4HAAAg2HxJ\njTWHD5TdR4qOkWLjyu5TpMTkieTk5Hr16qWnp996661yvj0pKclkMu3cuTM1NbV4+5gxYzIz\nM3fs2CGzjCLlCHZr1qzZtGlTQcEfxjB37drF2sUAAKBakFq2EbZuEvLL2uHX17lbhY+fkJAg\nCMKJEydkBjuNRvPkk0/OmjXr0UcfLdoz9/Tp0xs3bpw6dWoFCpAb7BYvXjxu3Dir1erxeBwO\nR8OGDb1e74ULF2rXrv3mm29W4MQAAAAKk3R6z32DdZ+8f8MOjZt60zrezCmioqIuXbpUun3v\n3r1r166dOHFi8QmwgiC88MILX331VVpa2uTJk2+99dajR4++/vrrdevWffrppytwdrnLnSxY\nsCAlJSUzM/PEiRMajWbbtm3nzp3bvn27x+NhRgUAAKgufElNPY8NF0zm67zVso3noT8JGrnp\n6LqSk5OXLVtWuv3AgQMzZszIyckp0R4ZGblv375BgwYtWLCgf//+c+fO7dev37///e8SK5DI\nJEqSJKdfVFTUn//851mzZgmCkJaWNm3atMGDBwuCMG7cOLvd/sknn1Tg3DfD4XBUbEXm6ig6\nOtput4fJCoIGg8FqtYbV9Y2Li8vOzg51FQoxGo2RkZH5+flOpzPUtShBFMWYmJjSP8fVymKx\nmEym4vuMq5tOpzObzTZbWTf11MRqtRoMhuzsbJ/PV7lHLroFWbkcDkcZvzrFwkLN/t2ak78K\ntlxBb5DqJnhbt5UaNi7jgGazWavVBqHSyiT3VqzJZCq6kKmpqTt27PAHu06dOr344ovBqg4A\nACAIJKPR26Wbt0vFH6ermuQONiYnJ2/YsMHlcgmCkJKS8tVXX/nbjx8/Hj5/rAAAAFRlcoPd\nlClTfv7552bNml27dq1r164ZGRnjx4+fO3fu0qVLu3TpEtQSAQAAIIfcW7H33HPP8uXLV6xY\nIUlSamrqiy+++PLLL7vd7qSkpDlz5gS1RAAAAMghd/JEafn5+WfPnm3evLler6/cmuQIq4fr\nmTyhbkyeUDEmT6gbkycqS0gmT1RAtZg8IfdWbN++fVeuXFl8deKoqKiWLVuGJNUBAACgNLm3\nYnfs2LFhwwar1frQQw8NGzasW7duxbe8BQAAqFI0N7ccXTUlN9hduXLlm2+++fTTT1etWrV0\n6dJGjRoNGzZs6NChFdieFgAAINiMRmOoSwgBuWHWbDY/+OCDq1evzszM/Oyzzzp16vTmm282\na9asa9euixcvDmqJAAAA5eX1ej2VqsLTEpRU8ckTNpttypQpixYtkiSpUj6qJEkrVqzYsmWL\nz+fr2rXr8OHDy3hEMawermfyhLoxeULFmDyhbkyeqCxMnqhEcm/FFnE4HN9///0XX3yxfv36\nnJycmJiYgQMHVkopa9as+eabb/7yl7/odLr58+cLgjBq1KhKOTIAAEA4kBvscnJy1q9f/+WX\nX27YsMHhcFit1vvvv//hhx/u06ePwWC4+Tq8Xu8333wzbNgw/3LHo0aNWrhw4ZAhQ8LzBjkA\nAEAFyA12tWrV8ng8kZGRAwcOfPjhh/v27RsREVGJdZw/fz4nJyctLc3/Mi0tzeFwnDp16tZb\nb63EswAAAKiY3GA3ePDghx9+uF+/fiaTKRh1ZGdni6IYFxfnfxkZGRkREVH8wZTCwsLVq1cX\nvWzZsmVycnIwKqmCNBqN0Wis9Gcaqib/4wt6vT5I/9KqIFEUw+fD+le+NBgMYbIMgSiKYXV9\ndTqdIAgRERFhssSpRqPRarXhc339P5+NRmO1mEMQtuQGu1WrVgW1jvz8/IiIiOI/600mU/En\nUgsKCubNm1f0cuzYse3btw9qSVVK+Pzg8NPr9WHyi8HPYrGEugRFRUREVO6QfxUXbtc33H5e\nhdv1NZvNoS4BZSn35IkgsVgsTqdTkqSidY8LCgoiIyOLd5g1a1bRy/r16+fn5ytdZYiYzebC\nwsIwGbHT6XQmk8nlcoXJrElBECIjI+12e6irUIherzcajYWFhW63O9S1KMRisVy7di3UVSgk\nIiLCYDBU+lTEKkuj0URERBTfk0ndTCaTTqez2+2VPmIXFRVVuQeUSZJ8V+yHbIXn9RpzrajW\nZkNQJucqrKoEu9jYWEmScnNzY2NjBUEoKChwOp3+r/0MBsNdd91V9DKslsMwGo0ulytMflBK\nkmQymTweT/gEO/9fNaGuQiH+v9zcbneYfGRRFM1mc5h8WOH3W7Eulyt8ljvR6/Xhc339A+0u\nl6vSBxqUD3Y+ybv37IL/nnn7muuKv0UjapNq3NW92cvxllsULqZyVZXHXBo2bBgdHb1//37/\nywMHDphMpmbNmoW2KgAAoDIeb+HnBx7a8utzRalOEASf5D2ZteGj//Q4dXWj/EMNHTpULMZk\nMrVt23bNmjXlLcnlcs2YMaNr167R0dEtWrQYMmTI0aNHy3sQv6oS7LRa7T333PPxxx+fOHEi\nPT39/fff79OnT1g9hQMAABSw+cTk01c3Xfctt8/xr5+H5zhOyj9a586df/rdunXrmjVr9thj\nj+3du1f+ETIzMzt06DB//vwBAwasWbPm2WefzcvLS0tL+/jjj+UfpEgFb8Vu3759+fLlXq93\n1KhR3bt3r9hBSnj00Ufdbvfs2bN9Pt/tt98+cuTISjksAACAX6b9yMELy8vo4PLafzz58n2t\ny+pTXExMTKdOnYpe3nHHHdHR0Rs3bixawS2gyZMn5+fnHzx4sHbt2v6WESNGPPvss08++eSd\nd96ZkJAg8zh+FRmx++qrr3r27JmXl1dQUNCrV6+1a9dW4CCliaI4bNiwJUuWLFu27PHHHw+T\n1RAAAIBijl3+PGCf9Kxv3N4KznkyGAxGo7FGjRql31q6dKkoiidP/mE48Pz58x988MGrr75a\nlOr8XnjhhdjY2Lfffru8BVQkPD3//PNPPfXU559/vmbNmgkTJrzwwgsVOAgAAIDCMu2Bn13z\neAuzHekVOLjNZps9e7bRaOzXr1/pd9u0aTN16tSYmJjijYcOHZIkqXfv3iU6R0REdOvW7eDB\ng+WtoaxbsRcvXqxdu3bp/W5PnTo1bdo0/9edO3devHhxec8KAACgPI+3UFY3n6xugiB89913\nRSu1CYKg1Wq3bt1av3790j07dOjQoUOHEo0ZGRlms/m6I3yJiYl79uyRWUaRskbs1q5d27p1\n69J3WtPS0latWuX1er1e78qVK8NqoWAAAFB9RZsSZXUzyuom/HHyxKZNm+69997Bgwe7XC6Z\n396oUSOHw5GdnV36rbNnzyYlJck8TpGyRuyeeOKJ2NjYZ555Zvbs2a+99lq3bt387XPnzu3V\nq1fTpk0FQbDZbJs3by7vWQEAAJTXOP7usidPCIJQOyolMqKuzAOWmDyRnJxcr1699PR0mZvd\nt27dWhTFTZs2Pfzww/4Wt9vtcDhMJtPOnTsfeeQRmWUUKWvEThTFxx577JdffnnssccGDx58\n3333HT58WBCEtm3bHjlyZMqUKZMnTz58+HDbtm3Le1YAAADlNY3vVzuqTdl9bms8pcLH909i\nPXHihMz+9evXHz58+LPPPpuZmelv8d8Lfeqpp3JyciZMmFDeAgJPnjAYDE899dTJkyfbtm17\n++23jxo16ty5c3Xq1Bk/fvwTTzxRt67cSAsAABBaoqgZ0Pp9oz72Rh1S649tVvPemzlFVFTU\npUuXSrfv3bt3+vTppe+6zp49OzIyMiUl5a233vrhhx98Pp/T6Vy8ePGQIUPq1KlT3rPLnRUb\nFRX10ksvHT9+3Gg03nrrrZMmTbru/WAAAICqLM7cbGiHH+paS84Q0GvMPZq9fNctr9/k8ZOT\nk5ctW1a6/cCBAzNmzMjJySnRXrNmzd27d48bN+7TTz8dOHDgs88+m5aWNn/+/NWrV8+fP7+8\nZxcDbuW7ffv2pUuX5ubm9u/ff8yYMaIo/vrrr88999ymTZumTJny17/+1WQylfesNy+s9oqN\njo622+1hsleswWCwWq1hdX3j4uLC588ko9EYGRmZn58fJttriqIYExNT+ue4WlksFpPJlJub\nGz57xZrNZpvNFupCFGK1Wg0GQ3Z2dqXvFRsfH1+5B/RzOBxl/OqUBOls9vbTVzflOy/oNMba\nUW1b1LrfElH7Rv0FQTCbzaWXCgme9PR0SZLKu71qgBG7L7/8smfPngUFBUlJSZMmTRo/frwg\nCM2aNVuzZs2GDRs2bNjQrFmzpUuXVrxqAAAAxYmC2DCue49mLw9otazfrQvaNRhbdqpTXtOm\nTcub6oSAwe4f//jHX//61zVr1sydO/fLL79cvHjx2bNn/W916NDhhx9+WLJkyTvvvFORegEA\nAFCpAgS7jIyMjh07+r/2r1d35syZ4h369u27f//+IBUHAAAA+QIEuy5durz//vuFhYWCICxY\nsMBisbRpU3KSMJu6AgAAVAVlLVAsCMK8efPuuuuuBg0axMTEnDt3bsmSJdHR0cpUBgAAgHIJ\nEOyaN29++PDhb7/91maz9erVq3HjxsqUBQAAgPIKEOwEQbBarY888ojL5TIYDP6WCxcu7Nmz\nx2KxtG/fPiYmJsgVAgAAlJter9fpAucc+arFs2eBS1y9enWLFi3ef/99/8u///3vDRs2HDhw\nYO/evRMTE1nrBAAAVEF6vd5QqURRDPVnCixAkl2zZs2jjz7avHnzVq1aCYKwdOnSV155pX//\n/sOHD/d6ve++++7o0aOTkpLuvPNORaoFAACQxe12V+5aynq9vuoP2gXYeaJt27Y6nW7Hjh1G\no1EQhLS0tIiIiJ07d/pDq8fj6dChQ40aNTZt2qRQvb8Lq50J2HlC3dh5QsXYeULd2HmisoRk\n54kKUHjniYoJEDxPnDgxcOBAf6rzv+zXr1/RUKROp3vggQdYxw4AAKAqCBDs6tSpc+TIkaKX\nzZo1K/GnZ25ublxcXFBKAwAAQHkECHYjRoxYs2bN/Pnz/eOuw4cP/+ijjy5duuR/NyMjY/ny\n5bfffnvQywQAAEAgASZPTJ069T//+c9TTz316quv9u3bt379+kajMSUlZdCgQS6X64svvrBY\nLP/4xz+UqRUAAABlCBDsDAbDunXrvvnmm4ULF3788cdut9vfvnjx4piYmCeeeOLpp5+uVatW\n8OsEAABAAIEX7tNoNPfee++9994rSVJ2dvZvv/0mSVJCQgKP1gEAgOrrrMu9Oc92we0xa8S2\nZnPXSItBUw1WqiubrOVYMjMzv/32223btkVFRbVs2bJVq1ZFqe7EiRObN28OZoUAAACVKdvj\nffz02VY///K3sxdmX7r8woXfBv56qu2RY1/l5JXrOJIkLV68OC0tLSoqqkaNGnffffeOHTvK\nW0xhYeG0adPatWtnsVgaN278xBNPXL58ubwHKRI42M2cOTMhIeGee+7p2bNn06ZNd+3aVfzd\nhQsX3nXXXRU+PQAAgJKuuD13Hvv10+xc4Y/Dc+dd7qGnzsy/nCX/ULNnz544ceKwYcO2bt36\nxRdfNGzYsGfPnj/99JP8IxQUFHTs2HH16tWjR4/+17/+NW3atK1bt/bu3bvCi7kGuBW7cuXK\nadOmde/efcSIEWfOnFm2bFmfPn327dvXvHnzip0PAAAghB7POHvK6brRu9POX2xnMd0WaZFz\nqMWLF0+aNOlvf/ub/2X37t3T09M//PDDzp07yyzmtddes9ls+/btK7oXev/99zdu3HjFihWj\nR4+WeZDiAgS7t956KzU1dfPmzf6llocOHdqlS5dRo0ZVYKQRAAAgtLbn27fZ7GX3mXHx8jfN\nG8s5Wl5e3okTJ4q3LFy4sKCgoHTPpUuXjh49Oj09vUmTJkWNkiS9+eabs2bNKj5voWbNmu+/\n/77VapVTQGkBbsUeO3bsgQceKNpAo3HjxkuWLNm5c+dnn31WsfMBAACEyjoZT9Htyrdflbct\n3vjx41euXNm6des33njjwIEDkiS1aNGibdu2pXu2adNm6tSpMTExxRszMjLsdvttt91WovOD\nDz7Yp08fOQWUFiDY1a5du2g5Yr8BAwb06dPnmWeeCZ/dDwEAgDqUcRO2iE8QTsvoJgjCjBkz\nvvnmm7S0tLfffjs1NbVWrVpPP/203X6dEcEOHTrMnDmzRo0axRszMjIEQahbt66s0uUJEOza\nt2+/evXq4ruKCYKwcOHCrKys4cOHS5JUiaUAAAAElShvPRNRkLvuSb9+/ZYvX37u3Lnjx4+P\nHDlyyZIlDz30kMzvTUxMFAShxAiaIAhZWVnp6ekyD1JCgGA3YcKE/Pz8Tp069enT5+uvv/Y3\nNm7ceP78+evXr09OTt65c2fFTgwAAKCwZhERAfvoRLGJ0RCw25EjRx566KGivRuaN28+e/bs\nefPmfffdd/n5+XKKSUpKMplMpaPUmDFjRowYIecIpQUIdp06ddq2bVvnzp23bdtWfPruyJEj\nP/30U4/Hs3v37oqdGAAAQGEDY2MC9ukeZYn5fXZBGaxW6+eff75hw4bijTabLTY2NjIyUk4x\nGo3mySefnDVrVlbW/62xcvr06Y0bN/bt21fOEUoTZd5OlSTJ6XQajcbijT6f79ChQ+np6YMH\nD67Y6SvM4XBUeImXaic6Otput3u93lAXogSDwWC1WsPq+sbFxWVnZ4e6CoUYjcbIyMj8/Hyn\n0xnqWpQgimJMTEz4PJFssVhMJlNubq5H3oPn1Z1OpzObzTabLdSFKMRqtRoMhuzsbJ/PV7lH\njo+Pr9wD+jkcjuv+6nwkPePbvBtetQiN5vsWTVLNptJvmc1m7R8D39ixYz/55JOpU6d27tzZ\nYDDs2bPnxRdfnDJlynPPPVfie/fu3bt27dqJEyeW2LjLbrenpqa6XK7JkyffeuutR48eff31\n1/V6/cGDB81mczk+7e8CbynmJ4piiVQnCIJGo0lJSUlJSanAiQEAAJT3bqMG/X89ddhxnUVJ\nBEF4o0HCdVPddS1YsKBFixbr1q2bO3euJElNmzadM2fO448/XrrngQMHZsyYMWLEiBLBLjIy\nct++fdOnT1+wYMHp06fr1avXr1+/l19+uWKpTpA/YlfVhNWIDiN26saInYoxYqdujNhVFoVH\n7ARBuObzPX/ht+WZV93FUlBzY8TrDRJ6WqNudMDSI3ZVkNwROwAAAHWwaDRvNkh4rm7tLbb8\ni263SaNpaza1M5s1cufCVl0EOwAAEI7idNrBcYHnUlQvAWbFAgAAoLog2AEAAKiE3GBXt27d\nNm3alG5v165d5W6FAQAAgIqR+4xdVlbWdWeCZGdnF19VDwAAAKEiN9gV7ZhRgn//WgAAAIQc\ns2IBAIAKGY3Gyl2sV6OpBjMTZAW7zMzMPXv2mEym2267zWD4w7a4J06cOHfuXK9evYJTHgAA\nIeYt0BRe0nkLRZ1FMtZxayKq5cL+Yaha5LBKFzjYzZw58/nnn/cvI96gQYNVq1bddtttRe8u\nXLjQv41GEGsEACAUPPnazO1mx6kI6fetFkStEJVcWOO2a1oTv/iqOqfTWbmbNkVERFT7nSdW\nrlw5bdq07t27jxgx4syZM8uWLevTp8++ffuaN2+uTH0AAISE84ru4lqrt+APoz6SV7AdNjrO\n6usNztPF3ehbUSV4vd4w2Y2zuADB7q233kpNTd28ebM/og4dOrRLly6jRo3asWOHIuUBABAC\nPqd4aV3JVFfEY9P+tt7a6P/ZFa4KCCjA7edjx4498MADRQOPjRs3XrJkyc6dOz/77LPg1wYA\nQGjk7jd57GX9inRm6mxHDWV0AEIiQLCrXbv2pUuXircMGDCgT58+zzzzTE5OTjALAwAgZOwn\nIgL2sR0n2KHKCRDs2rdvv3r16iNHjhRvXLhwYVZW1vDhw5kzAQBQH8kruHIDPSMvCc6scJx0\niSouwD/KCRMm5Ofnd+rUqU+fPl9//bW/sXHjxvPnz1+/fn1ycvLOnTuDXyQAAAqSRCHgwIUo\nCD5RiWKA8ggQ7Dp16rRt27bOnTtv27btp59+KmofOXLkp59+6vF4du/eHeQKVSu/8PzxK58f\nuPjesSuf5ReeD3U5AID/EXWS1uIL2E0XHXYzLtVE8gm2E7qLG4xnPjOd+8qU+ZPBnaeGIdjA\n69h17tx506ZNkiQ5nc7i7YMHDx40aNChQ4fS09ODVp462Zxnt6Y/m5G9Ufr9T0JREBvF9e7R\n5FWrsWFISwMACIIgRDZ25R0ylt0nqqkn4PgIqqbCK5rz60zOq/93+WzHdZk7I+I7uGp2dYqy\nr6okSe+9996iRYtOnDhhMBjat28/ffr0rl27lqsYl8s1e/bs77777tChOsOJeAAAIABJREFU\nQ3Xq1ElLS/v73/9+6623lusgReTW7na7jcb//RO/ePHi7t27vV6vRqNJSUkZPHhwxc4dnjKv\nHV65r/fp7O+lYgP9kiCdzv5+1YE+V+w/h7A2AIBfbPsCUVfW7VityReb4iyjA6qsgsua0yvM\nxVOdn+QVMn8yXPwuQKAvbvbs2RMnThw2bNjWrVu/+OKLhg0b9uzZs/gdzoAyMzM7dOgwf/78\nAQMGrFmz5tlnn83Ly0tLS/v444/lH6S4AMHO4/HMnDkzOTl55cqVRY2HDh3q2LFjbGzs3//+\nd5fLVbEThye3z7Hu6NBCT/Z13y1wZ68/OsztvaZwVQCAEnRWb627brhMnaiT6tyTz95i1ZHk\nEy6sN/lcN3w+MvewPu8XvcyjLV68eNKkSX/729/S0tK6d+++ePHibt26ffjhh/LrmTx5cn5+\n/sGDB6dMmXL33XePGDHi66+/njBhwpNPPnnx4kX5xylSVrBzu929e/eeNm3atWvXGjb8v1uE\nLVu2HD9+fExMzCuvvNKzZ0+fL/CDCPA7eHFp2Y/T5TsvHLj4nmL1AABuJKqFM2GgTW8t+TvO\nUMNT78E8U313SKrCTbKd0JUeqysh6ye5C9nk5eWdOHGieMvChQvHjh1buufSpUtFUTx58mTx\nxvPnz3/wwQevvvpq7dq1i7e/8MILsbGxb7/9tswyiivrs7377rtbt26dNGlSRkZGjx49itrr\n16+/cOHCjIyMZ555ZteuXYsWLarAicNTeta/Avb5VUYfAIACzA1dicNy6t5ri00rsLYqjO3g\nSBiU12BIrrG2J9SloYKunQ48u6AwU+O2yZryPH78+JUrV7Zu3fqNN944cOCAJEktWrRo27Zt\n6Z5t2rSZOnVqTExM8cZDhw5JktS7d+8SnSMiIrp163bw4EE5NZRQVrD76KOPmjZt+uqrr2o0\n1+mm0Whef/31xMTEVatWVeDE4Snb8WvAPjky+gAAlCFqJUsTV42u12r1ste4zWFOdMt/sh5V\nkEteYnPny7rMM2bM+Oabb9LS0t5+++3U1NRatWo9/fTTdvt1buJ36NBh5syZNWrUKN6YkZFh\nNptLNPolJiZmZGTIqaGEsuo+efJkt27ddLobZltRFO+4447jx49X4MThSZICz42XJJ8UeAEl\nAABQbhp5d1nLnjpTXL9+/ZYvX37u3Lnjx4+PHDlyyZIlDz30kMzvbdSokcPhyM6+zpP3Z8+e\nTUpKknmc4soKdhqNxmQylf39kZGRzJ+QL9oY+CLFmBuLAoteAgBQ+Uy1Ak8MEHVSRFzgbkeO\nHHnooYfc7v89bdm8efPZs2fPmzfvu+++y8/Pl1NM69atRVHctGlTUYvb7c7Ly3O5XDt37kxJ\nSZFzkBLKCnatW7fes2dP2d//008/JScnV+DE4alpfP+AfRrH9VOgEgAAwpA12S0G2i7O2sKj\nkTEv1mq1fv755xs2bCjeaLPZYmNjIyMj5RRTv3794cOHP/vss5mZmf6WlStXtm/f/qmnnsrJ\nyZkwYYKcg5RQVrDr2bPn7t27P//88xt1WLZs2YEDB7p3716BE4entgljTfq4MjqY9HGp9cYr\nVg8AAGElIs4Xl1rWnUatUardTdatyAYNGowePfqRRx55+eWXN27cuG3btjfffPO555575pln\nRLHknbe9e/dOnz699F3X2bNnR0ZGpqSkvPXWWz/88IPP53M6nYsXLx4yZEidOnXK9dH8REm6\n4V1kp9PZvn37U6dOvfTSS+PGjSseP2022zvvvDNjxoyEhIRDhw5ZLJYKnPtmOBwOh8Oh8Ekr\nxfncHV8dGeLxFZR+S6cx3dfykwYx3Uq0R0dH2+12rzcs9q4xGAxWq7X6Xt8KiIuLu+4DFqpk\nNBojIyPz8/NL7GSjVqIoxsTE5OTkhLoQhVgsFpPJlJub6/GExaRRnU5nNpttNluoC1GI1Wo1\nGAzZ2dmVvsxZfHx85R7Qz+FwXPdXp+QVLnxtyjumk4SSTz5pjVKDQQWWBtf/hWs2m7XaPwz3\neTyet99+e926df75rU2bNh07duzjjz9eOtj9//buPD6q6v7/+LlzZ89kJgnZWBIWCQQR0AAF\nRDEoRRAX0IKgIpWyWcX2q/0KKlWq9otUBfWn4gqo6Dcii+0XsVatLa24sIkgxYDIphCWLDOT\nSSaz3N8f105jgCwkmcmceT3/4JG599yZz3BzZ945995zXn755WnTpu3du/ecc86ps6qmpmbB\nggXvvffezp07k5KSBg8ePGLEiPvvv/93v/vd7bff3sQ3XW+wE0IUFxePGzdu165dLperT58+\nnTt3VhRl3759O3fudLvdXbt2Xbt27dmdA26muP7iP+b98sM9dx7z/ug25kxH38vyFmU6TvOf\nSbCTG8FOYgQ7uRHsWkqUg50QQmiibKfpxCfmmvIfTl0ajCI5L5B1Sc2pIxdGnBrsWs/evXs1\nTcvLy2vqhg2M5tKjR4+tW7e++uqrq1ev3r1798cff6xpmqqq/fv3nzhx4owZM6LfVyeBTEff\niRe8f9S9+buKT6qDpRZjaqeUC7Md/RXuoQcAIAoUkdonkNonUFNuCLgVg0lY0sMGUxsakqJ7\n9+5nt2HDw/RZLJbp06dPnz5dCOHz+TweT7t27eoZAwWNoQilvXNge+fAWBcCAEDiMqeEzSkN\nN4sjDeczr9f79ddf+/3+Xr16paam2u32KJQFAACApqrv3J+maQ888EC7du0GDBgwdOjQjIyM\ne++9N0GunAAAAIg79fXYvfzyyw8++GCHDh2uu+46RVFWrVq1YMGC9PT0O++8M2r1AQAAoJHq\n67FbsmRJZmbm9u3bn3rqqSeffHLHjh1ZWVkvvPBC1IoDAABA49UX7IqLi6+55prITchpaWnj\nxo375ptvolIYAAAAmqa+U7FerzczM7P2kqysrLZzjd2po//JLaHer6IoifZ+Y11ClOjvNHH2\nb+T9xrqQKEm0/atLqDcr4mr/WiyWln1CgyEORiVr4K7YOjuv7exLo9HocrliXUWUqKrqdDrr\nH0paGvrvmMViMZkaMVGfFAwGQ+L8MusfizabzWq1xrqWKEnA/etwOBLn8yqh9q8+Nm9ycnKs\nC2msqA0m3KbE63B0wWAwcWYmSMCZJ6qrqxNn/6alpZWXl8e6iijRZ57w+XwJNfNE4uxffeYJ\nj8fTds7ttKrEnHnC7XbHy8wTiamBYLdjx47XX3898vDLL78UQtReorvxxhtbvLI2ripQ6g+W\nW4wpNlNarGsBAAAQov65Yht/4jX6ve6xmks0FA7sPPrKl0eWlfqK9SXt7D37tp/aO3uyamit\nU4cJ2GPHXLGyYq5YuTFXrNzibq7YxFRfj11RUVHU6ogL1cHydbtu/q7ik9oLT/q+/uibOcUn\n/njVua9ajIlypQUAAGiD6gt2119/fdTqaPs0ob27e1qdVBfxXcXGd3dPv+a8NxXRVu4vAQAA\niSYObtxtI/Yc/+PBsr/X0+BA2Ud7T/xf1OoBAACog2DXWLtK3miwzVdH695WAgAAEDUEu8Yq\n8WxruI234TYAAACthGDXKJrQakLeBpvVBD2aSIhhOQEAQBtEsGsURShJ5qwGmyVZsrl5AgAA\nxArBrrFyUoY13MbVcBsAAIBWQrBrrIJOt9bfQDWYLug4KzrFAAAAnCpe54qNvnb2Xhd2uW/j\n/t9r4vRnWwfnzk1P6tUir/VdxcY9J/50snJ3WAum2Lqel3Ntx6RLW+SZAQCAxAh2TTAw59dG\ng/Xj/Q+Hwj+aDUk1WIZ2+e0FHWc2/yWqg+V/+fq2b0v/ElnyvfuzXSVFmcl9r8h/yWXt2vyX\nAAAAsiLYNc0FHWed027MVyWvf1/xaXWw1GpM6+ga0jvrxmRrp+Y/eSDsW7Pj2uPeHaeuOub5\n8s0vrph4/l+c1pzmvxAAAJASwa7JnNacIZ3ntsYzf3pg4WlTna4qcOLDvf817rxVrfHSAABA\nAtw80VYEQpVffr+0/jYHy/5+zLs9OvUAAIC4Q7BrK753fxYMVzfY7EDZ31q/FgAAEJcIdm2F\nt+ZI45p939qVAACAOEWwaytMhqTGNLOoya1dCQAAiFMEu7YiK7lfY5plOhrVDAAAJCCCXVvh\nsnbt4BxUfxu7KaNz2mXRqQcAAMQdgl0bcsk5v1cNlnoaDOv2kMlgj1o9AAAgvhDs2pBMR7/R\n+S+cKboN7fLbnpnXRbkkAAAQRwh2bcs57a6YVPDXHhljjQarvsSgqF3TL73+gvUDcu6IbW0A\nAKCNY+aJNifVds7o/BeD4eqK6gNChB3mjpntcrxebygUinVpAACgTSPYtVFGg7WdvWesqwAA\nAPGEU7EAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQI\ndgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAA\nkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgB\nAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiC\nYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAA\nIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJY/RfsqamZunSpdu2bauoqMjLy5s6dWrXrl2F\nEGvWrFm+fHmkmaqqa9eujX55AAAAcSoGwW7hwoXffvvtzJkzU1JSioqK5s+f/8wzzzgcjpKS\nkoKCgquvvlpvpihK9GsDAACIX9EOdidOnNi0adP9998/YMAAIcScOXMmT568efPmwsLCkpKS\n/Pz8goKCKJcEAAAgh2hfY+d2u7t3796zZ0/9ocVisVqt5eXlQoiSkpLs7Ozq6mqPxxPlqgAA\nACQQ7R67bt26LVq0KPJw06ZNFRUVvXv31jStpKRk3bp1ixcv1jQtJydn9uzZ+fn5kZY+n++J\nJ56IPLzwwgsHDx4c1dJjR1VVu92uaVqsC4kGg8EghDCbzfoPiUBRFIfDEesqokRVVSGE1Wo1\nmUyxriVKDAZD4uxfo9EohLDb7eFwONa1RIPBYDAajYm2f5OSkhLk+yhOxeAaO52mae+///7z\nzz9/5ZVX5uXlnTx50mAw9OrVa968ecFgcNmyZQ899NCzzz7rcrn09n6/f82aNZHN09PTCwsL\nY1N6LFgslliXEFVGo1H/BEkQVqs11iVElclkSpxgJxJv/5rN5liXEFWJtn8T7fso7iitnbs3\nbtz4yCOP6D8vWbKkY8eOQoiSkpLFixfv379/ypQpo0ePPnUrv98/efLkWbNmXXrppfqSUChU\nXFwcaZCcnJw4fyQ5HA6fz5cgfwHrf/5WV1dXV1fHupYocTqdbrc71lVEicVisdlsPp+vpqYm\n1rVEg6IoycnJibN/bTabxWLxer3BYDDWtUSDqqpWq7WysjLWhURJUlKSyWSqqKho8eSQkpLS\nsk+YyFq9U2TQoEFFRUX6zzabTQhRXFz829/+tm/fvs8//3ykQ64Oi8WSkZGhX3unU1W1V69e\nkYc+n8/n87Vm4W2IpmmhUCgUCsW6kGjQz8CGw+EE+WLQJc6b1TtiQ6FQgrxlRVE0TUuQNyuE\n0P/+DAaDifOWE2r/6nkuFAolSEdDnGr1y5j068N0iqKEQqEFCxYUFhbee++9tVPdli1bbrvt\ntsjftT6f79ixY7m5ua1dHgAAgDSifRnTtm3bSktL+/Xr99VXX0UWdujQoU+fPl6vd9GiRWPH\njjWZTEVFRTk5OQx9AgAA0HjRDnaHDx/WNC1y1Z1u5syZY8aMeeyxx1566aXHH39cVdWCgoK7\n7747ce6LBAAAaL5oB7uxY8eOHTv2tKsyMjLuueeeKNcDAAAgDbrEAAAAJEGwAwAAkATBDgAA\nQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJRHuAYgAJKOhWvXvNQbeqGtXkbGHO\nUYQl1jUBgIwIdgBakRZUjv8tyfMvqxb+YclJIRTV4eyjpl9UqagxLQ4ApEOwA9BatKDy/RpX\n1ZG6nzNaSFR8YQuUGttfXUG2A4AWxDV2AFrLiY/tp6a6CN9BU9kmezTrAQDpEewAtIqQz+De\nYauvhSbKttrCNUq0KgIA+RHsALSKyv0mLVRvC0VoAcV3yBSlggAgARDsALSKoLtRV88FK7jI\nDgBaDMEOQKto7F0Rqta6dQBAIiHYAWgVprRgY5pZ2tV/vhYA0AQEOwCtIik3oNoa6I0zOsLW\nDoHo1AMAiYBgB6BVKCYtbXBl/W3aXVSp8CEEAC2Hz1QArcXVt9rVr/pMa1MH+pJ7+qNZDwBI\nj5knALSijEKvNStQ+klSwPOfPyPNKeF2QyuTupPqAKCFEewAtK7kXv7kfL//uLGmVDWZTak5\n1nBSpb+GVAcALY9gB6D1KcKSGbRkBq1WJckhPJ5Y1wMAkuIaOwAAAEkQ7AAAACRBsAMAAJAE\nwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAA\nQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASxlgXEMcOlf/j29L3KqoPGBRTO3t+XvrV7ZLy\nY10UAABIXAS7s+H1f//u7pnfuz+NLNkr/u+zg4/2zr5p+DmPqAZLDGsDAAAJi2DXZJU1R9/c\nPtrr//7UVV8dXeGuPnhN7yLVYIp+YQAAIMFxjV2TvV/8q9OmOt2h8g1bv3smmvUAAADoCHZN\nc8z75YGyv9bfZsvhp0PhmujUAwAAEEGwa5oGU50Qwh+sOOrZHIViAAAAaiPYNY3Hf7gxzdzV\nh1q7EgAAgDoIdk1jNNga08xkTGrtSgAAAOog2DVNelLvxjTLSDqvtSsBAACog2DXNN3ajTKr\nyfW3ae8c6LJ2iUo5AAAA/0GwaxqrMWVIl7n1NFAV88VdH4xaPQAAABEEuyY7v8OMCzrOPO0q\n1WD5aY+n2jsHRLkkAAAAQbA7O8O6PTym1/L0pHMjSwyKmps6/Pp+7/bMvC6GhQEAgETGlGJn\nqXv6mO7pYzz+QxXVBw2KMc2eZzWmxbooAACQ0Ah2zZJsyUm25MS6CgBAWxH0GoJuVRg0c1rI\nYNZiXQ4SDsEOAIAW4N1rLvvc7j/+wxerogp7l5p2QyrN7UKxLQwJhWAHAEDzaOL435Mqtv9o\nBHstJCq/MfsOmLIu9zi6M4E4ooSbJwAAaJbyrbY6qS5CCyol7yVHuvGA1kawAwDg7IWqDKWf\n2etpoAWVExuYZxJRQrADAODsVX5jDgeU+ttUfWcKevjCRTTwewYAwNlr1GlWTfhPcDYW0UCw\nAwDg7IUDjWqmNdSrB7QIgh0AAGfPmBxuTDM1mUFPEA0EOwAAzl5Sl4a77AzWsDUrGIViAIId\nAABnz9o+YOvYQLZLHVCl8H2LqOAXDQCAZskc6VHtZzwha+8cSLmgKpr1IJER7AAAaBaTM5xz\nfYU1+zQnW53nVbe/yk13HaKGu68BAGguozPUaUK574C58ltzwG1QVGFuF0zu4WeiWEQZwe4H\nVYHSL48s3V/6gdt/yGiwpCf17plxbV76NQp/ZwEAGkMR9i419i5MC4tYItgJIcT+sg//vHum\nP1gRWeKuPrTv5J+/cL50Za9ldnNmDGsDAABoJLqjxOGKj9d9dXPtVBdxxP35mh0/C4Qqo18V\nAABAUyV6sAuFa97/+o6Qdsae85O+f31+aFE0SwIAxJgm/CXG8u3Wss029y5LgGleET8S/VTs\nt2V/cfsP1t/my++XDc6doxrM0SkJABBD1UdMx/6aVFN7aldFOLr7Mwor6xnTBGgj4jXYKYpi\nNLZA8d9VbGywTU3IU1r9VXvnwOa/3NlRFEVVVUVJiHkGDQaD/m+L7N94kThvVlVV/d8EecuK\norTUh1Vc0I/fuH6/nmLTkXeTwnXGLdGEd4+l+qip80SvyfmfbKd/Msf1+20S/WtIVVV9R6Nt\nitdfR4PBYLPZmv88NeGyxjQLiPIWebmzYzAYrFarpmmxKiCa9M8Lk8mUIEFWCKEoSgx/u6JM\nD3Zms1n/IREk1P7VI47FYjGZTLGu5Wz4y5QjfzbXTXX/FvQYjqxL7jm1Rvz7w0n/qztx9q9+\n2NpstgT5PopT8RrsQqGQz+dr/vOoIqkxzbSA1ePxNP/lzo7L5aqsrAyFEmIwJLPZbDKZ/H5/\ni+zfuJCWlhbD364os1qtDoejqqrK7/fHupZoUBQlJSUlcfZvUlKSzWarrKwMBtvivKi+/Wb3\nLov/mDFcoxgdYXvngKtvlTH5Pz1wx/6aHK53bjDfEeXI1hpHjx9+e41Go91uT5z963Q6zWaz\n1+sNh1v4lLTFYmnZJ0xkid6b2j654ROsqsGSmdwvCsUAAFpDuEY5ss75/R+d3j2WQIUaqjL4\njxvLNtsOvJrq2WXV22hh4d3X8LXUnj1cb402LdGD3TnpY2ym9Prb5GeONxns0akHANCytLA4\n+o6z8pvTBDItqJS87/DstgghQj5D2N/wFSCBskS5igBxKtGDnVl1FJ6zoJ4GDkuHC7vcG7V6\nAAAty73T6jtY3zV/x//mCFcn+rchpMGvsuiRMfbS7o/qo5nUuRzUae089ryVdlNGTAoDADRf\nxXZb/Zf6h/2K+yuLag8bzA3fE2BKTYjLnRG/CHZCCNGn/c9vuOCjXlkT7KY0fUk7e88Lu9x7\nU/8N7ew9Y1sbAOCshXyGmlK1wTOsvu9MikEkdWt4mldHd6aCRZsWr3fFtrg0e4+RPZ4RQgRC\nlQbFxHDEACCBUFWj+i9CPoMQIu0nPu9esxY8Yw60ZAQdeQlxQzfiFz12dZnUJFIdAMjBYGnU\nwByqVRNCmFJDWT/1KurpT8gak8LZYzwKX5to2+ixAwBIy5gUNiWHG5zs1dr+h/HrHD38alL4\n+EeOmpM/uvvV0b0mvdBrTGJKMbR1BDsAgLwU4exTfXJjfUNWKaqW3Ks68tDWMZB7Y1n1UVP1\nUWPYr6hJYXtuwOTingnEB4IdAEBmKRdUeYrNNSfO+H2XNthXewZYIYRQhLV9INKNB8QRLhYA\nAMhMMWodrnFbMk8/y1nqQF9q/6oolwS0HnrsAACSMzrCnSaUV+y0enZZ/ceNQhMGk2bvHEjp\n77Nmt8VpbYGzRrADAMhPUUVKv+qUftVaWGhBpTFjEQPxiGAHAEggikEopDrIi2vsAAAAJEGw\nAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJMEDxWaqo/varo28c\ncW/yh9w2U7tOrqG9s2+0mzJiXRcAAEhcBLsm04T2+cHHNx1aHArXRBYeLPvbpkNPFJ6z4Nys\nSTGsDQAAJDKCXZNt/PahzYf/36nLA6HK94vvCGvB87InR78qAAAArrFrmiPuz0+b6iL+9s09\n7upDUasHAAAggmDXNFsOP1N/g1DYv/3IS9EpBgAAoDaCXdMcLP97w23KGm4DAADQ4gh2TVAT\n8gRClQ028/qPRKEYAACAOgh2TWA02A2K2mAzi9EZhWIAAADqINg1gUFRMxx9G2yWnVwQhWIA\nAADqINg1TWOGqWMoOwAAEBMEu6bpnXVT/R1yeelX56YWRqscAACA/yDYNY1qMF117muZjn6n\nXds59bKf9qxvlDsAAIDWQ7BrMrs5c0K/d4Z2vd9p7RxZmJ507k97PHlN7zdMBnsMawMAAImM\nKcXOhmqwDOg0e0Cn2b6aY1XBMrspw2ZKi3VRAAAg0RHsmsVuzrSbM2NdBQAAgBCcigUAAJAG\nwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAA\nQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQhDHWBQAAEA1B\nryFcbVDtYdUejnUtQGsh2AEAZKYFlbJtVvcOa9Cj6kvMqSFnvyrXedWKGtvSgJZHsAMASCvk\nM3z/R6f/2I++7GrK1BN/c1TutbS/0m2waLGqDWgNXGMHAJCTFhZH/q9uqouoOmw6+l5ylEsC\nWhvBDgAgJ88ua/XR+k5M+b41V+4zR60eIAoIdgAAOXl2WVukDRBHCHYAABlpovpYwzdHVJdw\nrTmkQrADAEhICytaSGmwWTjQcBsgjhDsAAASUlTNYG14vDqjgzHtIBWCHQBATkmdAw200IQ9\ntyYqtQBRQrADAMgppaCq/gaKUUu5oDo6xQDRQbADAMjJkhlMG+wTQogzDEKcUVhpTA5FsySg\ntXE3EABAWmmDfAaTdnKjvc6NFAazln6J13muP1aFAa2EYAcAkFlKQVVSnt/7L2vVd8ZQtUFN\nCts6BJy9/aqN2yYgIYIdAEBypuRw6k98qbEuA4gCrrEDAACQBMEOAABAEgQ7AAAASRDsAAAA\nJBGDmyfWrFmzfPnyyENVVdeuXSuE0DTtjTfe+Oijj8Lh8EUXXTRlyhRVbXj+ZgAAAOhiEOxK\nSkoKCgquvvpq/aGi/DC20MqVK9evX3/77bcbjcann35aCDF16tTolwcAABCnYhPs8vPzCwoK\nai8MhULr16+/+eabhwwZIoSYOnXqkiVLbrjhBqvVGv0KAQAA4lEMrrErKSnJzs6urq72eDyR\nhYcPHy4rK+vfv7/+sH///j6fb9++fdEvDwAAIE5Fu8dO07SSkpJ169YtXrxY07ScnJzZs2fn\n5+eXlpYqipKWlqY3czgcFoulrKwssqHX67377rsjD0ePHj1q1KgoFx8rRqPR4XDEuooo0U/N\nW61Wk8kU61qiRFEUl8sV6yqixGAwCCHsdnvidMYbDIbE2b/6hdGhRi80AAAVEUlEQVQOh0PT\nzjA5q1wURUnA/ZucnBzrQlCfaAe70tJSg8HQq1evefPmBYPBZcuWPfTQQ88++6zH47FYLPqH\nvs5ms7nd7sjDQCDw+eefRx6ef/75ifPFL4RIqDcrhDAYDLV/GaSXaPtXVdWEujUq0fav0ZhY\ncxol1IeVSLzf57jT6offxo0bH3nkEf3nJUuWdOzYcdWqVZG1d9xxx+TJk7ds2eJyufx+v6Zp\nkXspqqqqandTpaambt68OfLQ5/OdOHGitYtvI1wul9frDYVCsS4kGsxms9Pp9Pl8Pp8v1rVE\nSVpaWmlpaayriBKr1epwODwej9+fEJOvK4qSkpJS++SD3JKSkmw2W3l5eTAYjHUt0WA0Gu12\ne+0+CLk5nU6z2VxaWhoOt/A0u+np6S37hIms1YPdoEGDioqK9J9tNludtRaLJSMjo7y8vEuX\nLpqmlZeXp6amCiGqqqr8fr/+MwAAABqj1TuQVVW1/5uiKFu2bLntttsif9/4fL5jx47l5uZ2\n7tzZ5XJt27ZNX/7FF1/YbLa8vLzWLg8AAEAa0b4Sok+fPl6vd9GiRWPHjjWZTEVFRTk5OQUF\nBQaD4YorrlixYkWnTp0MBsOyZctGjhxpsViiXB4AAED8inawM5vNjz322EsvvfT444+rqlpQ\nUHD33XfrV55OnDgxEAj84Q9/CIfDQ4cOveWWW6JcGwAAQFxT4vSm9IS6uJ6bJ+TGzRMS4+YJ\nuXHzREvh5okWlFg3aQMAAEiMYAcAACAJgh0AAIAkCHYAAACSINgBAABIIrFm9ItTwWAwTm9e\nPgvl5eU7d+5MT09PS0uLdS1REggEYl1C9Bw7duzLL7/Mzs52Op2xriVKEmr/HjhwoLS0NDc3\n12q1xrqWaNA0LUHu/9Xt3r3b6/V269Yt0aYDji/xOtwJZPXJJ5/Mnj17xowZM2bMiHUtaHkr\nV678wx/+8OCDD15xxRWxrgUtb9GiRW+88cbSpUv79u0b61rQ8u68884NGza8//77TPjZlnEq\nFgAAQBIEOwAAAEkQ7AAAACTBNXZoW06ePLlt27Zu3bp169Yt1rWg5R06dOjrr7/u3bt3+/bt\nY10LWl5xcfHBgwcHDhzocrliXQta3vbt248fPz5s2DCz2RzrWnBGBDsAAABJcCoWAABAEgQ7\nAAAASTDGINqEmpqapUuXbtu2raKiIi8vb+rUqV27dhVCrFmzZvny5ZFmqqquXbs2ZlWiGTRN\ne+ONNz766KNwOHzRRRdNmTJFVdVYF4Vm4bCV25n2I8dyG0ewQ5uwcOHCb7/9dubMmSkpKUVF\nRfPnz3/mmWccDkdJSUlBQcHVV1+tN1MUJbZ14qytXLly/fr1t99+u9FofPrpp4UQU6dOjXVR\naBYOW7mdaT9yLLdxBDvE3okTJzZt2nT//fcPGDBACDFnzpzJkydv3ry5sLCwpKQkPz+/oKAg\n1jWiWUKh0Pr162+++eYhQ4YIIaZOnbpkyZIbbrghQSaekhKHrfROux85lts+rrFD7Lnd7u7d\nu/fs2VN/aLFYrFZreXm5EKKkpCQ7O7u6utrj8cS0RjTL4cOHy8rK+vfvrz/s37+/z+fbt29f\nbKtCc3DYSu+0+5Fjue2jxw6x161bt0WLFkUebtq0qaKionfv3pqmlZSUrFu3bvHixZqm5eTk\nzJ49Oz8/P4al4uyUlpYqipKWlqY/dDgcFoulrKwstlWhOThs5Xam/cix3PbRY4c2RNO0v/zl\nLwsXLrzyyivz8vJKS0sNBkOvXr1eeeWVpUuXdunS5aGHHqqoqIh1mWgyj8djsVgMhv984Nhs\nNrfbHcOS0FI4bKV0pv3Isdz20WOHGNi4ceMjjzyi/7xkyZKOHTsKIUpKShYvXrx///5p06aN\nHj1aCNGuXbtVq1ZFtrrjjjsmT568ZcuWSy+9NCZl46wlJSX5/X5N0yLXX1dVVTkcjthWhebj\nsJXVmfajy+XiWG7jCHaIgUGDBhUVFek/22w2IURxcfFvf/vbvn37Pv/882eajMhisWRkZOgX\n8SC+pKamappWXl6empoqhKiqqvL7/frPiF8ctokjsh+7dOnCsdzGcSoWMaCqqv3fFEUJhUIL\nFiwoLCy89957a389bNmy5bbbbot08vt8vmPHjuXm5saoapy9zp07u1yubdu26Q+/+OILm82W\nl5cX26rQHBy2cjvTfuRYbvvosUPsbdu2rbS0tF+/fl999VVkYYcOHfr06eP1ehctWjR27FiT\nyVRUVJSTk8MYCvFIVdUrrrhixYoVnTp1MhgMy5YtGzlypMViiXVdOHsctnI70340GAwcy22c\nomlarGtAonv77beXLl1aZ+HMmTPHjBlz/Pjxl156adeuXaqqFhQU3HLLLcnJyTEpEs2kadpr\nr722YcOGcDg8dOjQW265pfb114g7HLbSO9N+5Fhu4wh2AAAAkiBlAwAASIJgBwAAIAmCHQAA\ngCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgB8erNN99UfszpdA4cOPDll1+u\nM/C43+9/+OGHL7vssszMzPT09EGDBj3zzDOBQOC0T7tjx44TJ05E5R3UR9O0Z5999vvvv4/O\nyw0cOPCnP/1pdF4LAFoPwQ6Ib9dcc828efPmzZt33333XX/99QcOHJg2bdr8+fMjDYqLiwcM\nGHD//fd7PJ7x48dfddVVpaWlt99++8iRI0OhUJ1n8/v9F1100YcfftikGtatW3fRRRelpaVl\nZ2dffvnlGzdurL32s88+Gz16dGZmZteuXW+88cbDhw9HVu3Zs2f06NEul6tfv35vv/127a2K\nioqefvrprKysJlVy1hwOR1JSUnReCwBakQYgPhUVFQkhVqxYUXvh8ePH27dvbzabPR6PpmnV\n1dW9e/e22+0rV66MtAkEAv/93/8thFiwYEHtbf1+/x133CGEKCoqamoZvXv3vu++++bNm5eb\nm2s0Gj/55BN97dq1aw0GQ35+/gMPPHD33XdnZWV16tRp3759mqa53e6cnJwRI0a8/vrrv/rV\nrwwGw4YNGyKVdOvWbe3atWf1HwMAiYtgB8Sr0wY7TdNmz54thNi8ebOmaS+88IIQ4tFHH63T\nJhQK5eTk5Ofn6w9LSkpGjRqVnp6u/73XpGDXvXv3/Px8PUdqmvbdd9/ZbLZx48Zpmub3+zMz\nM/v06VNVVaWvPXr0aEZGxi9+8Qu9/vT09Miq8ePHT5kyRf/5qaeeGjJkSONraI6ysrKz2zAU\nCgUCgZYtBgCaiVOxgJz8fr8QYvXq1S6Xa9asWXXWGgyGV199de7cufqVdnqn2k033XTttdee\n9tlMJlP79u1PXe71er/55pvLLrvM4XDoSzp06JCXl7dv3z4hxFdffXXs2LFp06ZZrVZ9bVZW\n1uTJk1977bXKykq3252SkhJZ1aFDh7KyMiGEx+N56KGHFi5cWP8bvP76681ms76JzufzORyO\n0aNH6w/Xr19fWFiYlZXldDoLCgpefPHFSMvhw4ePHz9+z549o0aNGjBggBBiyJAhta+xq2fb\n0aNHjxs37s0332zfvr3+3zJz5ky32x1p8Omnn+opuUePHr/4xS9OnjwZWbV///5JkyZ17drV\n5XJdcskl77zzTv3vEQCaimAHSKW0tHT16tWqqvbq1UsIsXv37vPOOy+SumorLCycMmWKyWQS\nQqSnpy9evHjx4sW//OUvT/u0WVlZGRkZpy632WxHjhypHcJKSkr27NkzZMgQIYTX6xVCqKpa\nexOr1VpTU3Po0KFLL7304MGDCxcu9Hg8H3300fLly6+44gohxKOPPjp48OCLL764/nd6/fXX\nBwKBdevWRZasX7++srLy5ptvFkK88sorY8aMKS0tnTJlyq233hoOh2fMmLFy5cpI44qKinHj\nxmVlZc2dO7fOMze47fbt23/+85+PHz/+xRdfvPjii1944YW77rpLX/XOO+8MGzbs8OHDt99+\n+1VXXbVq1aoBAwbo6XPHjh3nn3/+P/7xj4kTJ951110VFRVXXXWV3qUKAC0m1l2GAM6Sfir2\n2muvnT9//vz58x944IFbb701OztbCHHPPfdomlZZWakoyo033tj45/zggw9EE0/FRjzwwAM3\n3HBDdnZ2fn7+N998o2laRUWF1Wq98MILg8Gg3qaioqJbt25CiA8++EDTtNdeey1yy8L06dPD\n4fCRI0ecTueOHTsafDm9f04/56ubMGGC0+n0+Xyapo0cOTI3N9fv9+ur/H6/0+mcPn26/rCw\nsFAI8cILL0S2HTx48IgRI/Sf69921KhRQogXX3wxsu2AAQNyc3M1TQsEAj169OjTp4/X69VX\nvf/++0KIJ554QtO04cOHd+7cOXLmNxAIXHLJJUlJSW63u9H/xwDQAGNs4iSAFrJmzZo1a9bo\nP6uq2q1bt7vuuuvXv/61EEK/7k1RlOhU8te//vXQoUMlJSXnn3++pmlCCKfT+T//8z933nnn\n4MGDx44dW1lZuWLFinA4LITQq7rpppuuvPLKL7/8snPnzp07dxZCPPjgg9ddd915553X4MvZ\nbLarr7767bffrqqqstlsVVVV77zzzsSJE202mxBi9erViqKYzWa98YkTJ4LBYFVVVWTzpKSk\nqVOnnvaZG9zW4XDccsstkYd9+/b905/+JITYunVrcXHxSy+9FEmrI0aMePbZZ88999yysrKP\nPvro4YcfTklJ0VcZjcZZs2ZNmjTp008/ZaQVAC2FU7FAfKt980QwGCwuLv7Nb35jNBqFEO3a\ntVNVVb/c7VSHDx9+7733WnCguA0bNnz77bc7duz44osvRowYoV+991//9V8rV65UVXXhwoV/\n+tOfZs2a9Zvf/EYIofcsCiFSUlKGDRump7o9e/a89tprv/vd74QQ27dvv/TSS51O5wUXXKDH\nplNNmDDB5/O999574sfnYYUQDoejuLj46aefnj59ut6j5vP5am+bm5tb5xxxRIPbdu7cufa2\nBsMPH6R79+4VQvTu3bt241tvvfWSSy75+uuvhRDz5s2rPe7gpEmThBBtYdRAANIg2AHSMhqN\n/fv337lzp8fjOXXt73//+1GjRn333XfNeYmqqqqSkpJgMBhZ0rt371tvvXX//v07duzQl4wf\nP/7TTz91u907d+689957jxw5ImoFu9ruu+++WbNm5eTklJSUXHbZZS6X64UXXhg2bNi4ceP+\n8Y9/nNp+1KhRTqdT77B86623unTpErky7+GHH+7fv/9jjz2WlJQ0d+7cvXv36tkxop5R6xrc\nVr8w8VQ1NTVCCD1V16H3/82bN+9vpxg+fPiZKgGApuJULCCziRMn3nnnnU8++eS8efNqLw+H\nw++8805ycnJBQUFznv+f//znyJEj161bN2bMmMhCi8Ui/p1+li1b1rVrV/2aNt369esLCgrS\n0tLqPNWmTZs++OADvdNr7dq1ycnJq1atUlV14sSJhw4dWrZs2am3U1gslmuuuWbdunVut3vd\nunV33XWXfobX4/E8+OCDM2bMeO655yKNTx2N+bSas2337t2FELt379bvtNX9+te/7tmz5w03\n3CCEMBqNl1xySWTV119/vXnz5tqNAaCZ6LEDZDZ9+vQuXbo8/PDDy5cvjyzUNG3u3LmHDh26\n4447znQ6spGGDh1qNptfeeWVyJKampr//d//dblc+m25L7/88uTJkyNdhqtWrdq5c+ett956\n6lPNmTNnzpw5euDTfjwlmslkOlO0mjBhQllZ2Zw5cyorKydPnqwvPHDgQCAQyM/PjzT75z//\n2ci+yeZs279//w4dOjzxxBP6WDNCiI8//vjJJ5/0er0ul2vEiBHPPfdc5Mx4TU3NlClT7rnn\nHrvd3pgnB4DGoMcOkJnD4Xj77bfHjRt3yy23PPXUU4MGDaqsrNywYcOBAwcuvPDC+++/v5HP\n06VLl7S0tK1bt9ZZbrfbf/nLXz7xxBPDhw8fMWJETU3NW2+99a9//WvFihX6Gcn58+dffvnl\nQ4cOnTRp0p49e15//fWf/OQnN910U53n+fOf/1xcXBwZ123s2LHz5s2bMGHCpEmTPvvss9Wr\nV+u3655q5MiRKSkpzz///NChQ/UOMyFEz5499Th79OjRHj16fP7556tXr87Kyvrkk08+/PDD\nyy67rJ532pxtbTbb448/fuONNw4ePPhnP/uZz+d77rnncnNzZ8yYIYR49NFHhw0bpv9XZGdn\nv/XWW1u2bCkqKora3S0AEkKM7sYF0FxnmnniVF6vd+7cuUOGDElJScnMzLz44ouffvrpyBAk\ntZ1puBOj0ZidnX3aJw+FQk899VSfPn2sVmv79u1Hjhz597//vXaD9evXDxw40OFwDBw48L77\n7ovMUVH7Gfr161d7ABFN07Zu3VpYWOhwOPr27btmzZp63t3Pf/5zIcTzzz9fe+GuXbtGjRrl\ncrlycnImTZp06NChV199NTMz8/LLL9c0rbCwcMCAAbXb1x7upP5tR40apd/2GzFt2rT09PTI\nww8//HD48OEpKSkdO3a84YYbDhw4EFm1Z8+ea6+9tlOnTi6X6+KLL3733XfreV8AcBYU7cen\nPAAAABCnuMYOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMA\nAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATB\nDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABA\nEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsA\nAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ\n7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAA\nJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbAD\nAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAE\nwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASfx/oMijJ/+M7asAAAAASUVO\nRK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ntd <- normTransform(gdds)\n",
    "plotPCA(ntd, intgroup=c(\"exp\", \"cond\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential expression analyses\n",
    "\n",
    "## 1º Question: What are the DE genes for each time point and host phenotype between the control and inoculated groups?\n",
    "\n",
    "Here,I'll perform pairwise comparisons for each HPI and plant phenotype (e.g. Resistant at 24HPI; Susceptible at 24HPI, etc.). To this end, I create a new subset for each comparison from the main DESeqDataset. **In all of the following cases, the variable being compared is Control vs Inoculated**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### subset conditions ###\n",
    "pair_cond = list(\n",
    "    # Resistant at 24HPI\n",
    "    \"rh24c\" = colData(gdds)$hpi==\"24h\" & colData(gdds)$exp==\"R\",\n",
    "    # Resistant at 48HPI\n",
    "    \"rh48c\" = colData(gdds)$hpi==\"48h\" & colData(gdds)$exp==\"R\",\n",
    "    # Resistant at 72HPI\n",
    "    \"rh72c\" = colData(gdds)$hpi==\"72h\" & colData(gdds)$exp==\"R\",\n",
    "    # Susceptible at 24HPI\n",
    "    \"sh24c\" = colData(gdds)$hpi==\"24h\" & colData(gdds)$exp==\"S\",\n",
    "    # Susceptible at 48HPI\n",
    "    \"sh48c\" = colData(gdds)$hpi==\"48h\" & colData(gdds)$exp==\"S\",\n",
    "    # Susceptible at 72HPI\n",
    "    \"sh72c\" = colData(gdds)$hpi==\"72h\" & colData(gdds)$exp==\"S\",\n",
    "    ######\n",
    "    # Control at 24HPI\n",
    "    \"c24\" = colData(gdds)$cond==\"C\" & colData(gdds)$hpi==\"24h\",\n",
    "    # Control at 48HPI\n",
    "    \"c48\" = colData(gdds)$cond==\"C\" & colData(gdds)$hpi==\"48h\",\n",
    "    # Control at 72HPI\n",
    "    \"c72\" = colData(gdds)$cond==\"C\" & colData(gdds)$hpi==\"72h\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "getpair_stats <- function(ds, cond) {\n",
    "    pc <- unlist(pair_cond[cond])\n",
    "    sdd <- subset(ds, select=pc)\n",
    "    \n",
    "    if (cond %in% c(\"c24\", \"c48\", \"c72\")) {\n",
    "        design(sdd) <- formula(~ exp)\n",
    "    }\n",
    "    \n",
    "    sddd <- DESeq(sdd);\n",
    "    sddres <- results(sddd, alpha=0.05, lfcThreshold=1, altHypothesis=\"greaterAbs\")\n",
    "    \n",
    "    #print(sdd@colData)\n",
    "    print(paste(\"Data set with \", nrow(sdd@colData), \" samples and \", length(sdd@rowRanges), \" genes.\"))\n",
    "    #print(sddres)\n",
    "    print(paste(\"Total of significant DE :\", sum(sddres$padj <= 0.05, na.rm=T)))\n",
    "    \n",
    "    return(list(\"res\"=sddres, \"ds\"=sdd))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resistant at 24HPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Data set with  4  samples and  24413  genes.\"\n",
      "[1] \"Total of significant DE : 74\"\n"
     ]
    }
   ],
   "source": [
    "rh24 <- getpair_stats(gdds, \"rh24c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(rh24$res), file=\"rh24c.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ids = rownames(as.data.frame(subset(rh24$res, rh24$res$padj <= 0.05)))\n",
    "write(gene_ids, file=\"resistant24hpi.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resistant at 48HPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rh48 <- getpair_stats(gdds, \"rh48c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(rh48$res), file=\"rh48c.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ids = rownames(as.data.frame(subset(rh48$res, rh48$res$padj <= 0.05)))\n",
    "write(gene_ids, file=\"resistant48hpi.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resistant at 72HPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rh72 <- getpair_stats(gdds, \"rh72c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(rh72$res), file=\"rh72c.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ids = rownames(as.data.frame(subset(rh72$res, rh72$res$padj <= 0.05)))\n",
    "write(gene_ids, file=\"resistant72hpi.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Susceptible at 24HPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sh24 <- getpair_stats(gdds, \"sh24c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(sh24$res), file=\"sh24c.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ids = rownames(as.data.frame(subset(sh24$res, sh24$res$padj <= 0.05)))\n",
    "write(gene_ids, file=\"susceptible24hpi.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Susceptible at 48HPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sh48 <- getpair_stats(gdds, \"sh48c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(sh48$res), file=\"sh48c.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ids = rownames(as.data.frame(subset(sh48$res, sh48$res$padj <= 0.05)))\n",
    "write(gene_ids, file=\"susceptible48hpi.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Susceptible at 72HPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sh72 <- getpair_stats(gdds, \"sh72c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(sh72$res), file=\"sh72c.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ids = rownames(as.data.frame(subset(sh72$res, sh72$res$padj <= 0.05)))\n",
    "write(gene_ids, file=\"susceptible72hpi.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Question: What are the trends of DEG for each host phenotype across time points?\n",
    "\n",
    "First, a simple comparison of the significant DE genes for each plant phenotype across all three HPI points."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pairsr <- list(\n",
    "    \"rh24\" = rh24,\n",
    "    \"rh48\" = rh48,\n",
    "    \"rh72\" = rh72\n",
    ")\n",
    "\n",
    "pairss <- list(\n",
    "    \"sh24\" = sh24,\n",
    "    \"sh48\" = sh48,\n",
    "    \"sh72\" = sh72\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "x <- data.frame(\n",
    "    \"x\"=c(\"h24\", \"h48\", \"h72\"),\n",
    "    \"r\"=unlist(lapply(pairsr, function(x) sum(x$res$padj <= 0.05, na.rm=T))),\n",
    "    \"s\"=unlist(lapply(pairss, function(x) sum(x$res$padj <= 0.05, na.rm=T)))\n",
    ")\n",
    "x.m <- melt(x)\n",
    "\n",
    "p <- ggplot(x.m, aes(x, value)) +\n",
    "    geom_bar(aes(fill=variable), stat = \"identity\", position=\"dodge\") +\n",
    "    xlab(\"HPI\")+ ylab(\"Frequency\") + ggtitle (\"Number of DE genes for each phenotype across HPI\") +\n",
    "    scale_fill_discrete(name = \"Phenotype\")\n",
    "\n",
    "p\n",
    "x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When comparing the control with the inoculate group, we observed a substantical increase in the number of DEG along time for both host phenotypes. This increase, as well as the absolute number of DEG, is higher in the resistant phenotype. With the exception of the 48h time point, the number of DEG was between 2-3.5x higher than in the susceptible phenotype."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### 1.1.1 Question: What is the contribution of up and down regulated DEG for each host phenotype along time?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x <- data.frame(\n",
    "    \"x\"=c(\"h24\", \"h48\", \"h72\"),\n",
    "    \"rUP\"=unlist(lapply(pairsr, function(x) sum(subset(x$res, x$res$padj <= 0.05)$log2FoldChange >= 1, na.rm=T))),\n",
    "    \"rDOWN\"=unlist(lapply(pairsr, function(x) sum(subset(x$res, x$res$padj <= 0.05)$log2FoldChange <= -1, na.rm=T))),\n",
    "    \"sUP\"=unlist(lapply(pairss, function(x) sum(subset(x$res, x$res$padj <= 0.05)$log2FoldChange >= 1, na.rm=T))),\n",
    "    \"sDOWN\"=unlist(lapply(pairss, function(x) sum(subset(x$res, x$res$padj <= 0.05)$log2FoldChange <= -1, na.rm=T)))\n",
    ")\n",
    "x.m <- melt(x)\n",
    "\n",
    "p <- ggplot(x.m, aes(variable, value,fill=variable)) +\n",
    "    geom_bar(aes(), stat = \"identity\", position=\"dodge\") +\n",
    "    facet_wrap(~x, ncol=3) +\n",
    "    xlab(\"\")+ ylab(\"Frequency\") + ggtitle(\"Number of UP and DOWN regulated DE genes for each phenotype across HPI\") +\n",
    "    scale_fill_discrete(name = \"Phenotype\")\n",
    "\n",
    "p\n",
    "#x.m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, the majority of the DE genes are being upregulated, and they are the main contributors to the difference in total DEG between phenotypes across HPI points. This seems to lead to the conclusion that the response to the presence of the pathogen in either host results mostly in an increase of gene expression. As noted before, the number of up-regulated DEG is much higher in the resistant phenotype and seems to be elicited earlier."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2.1 Question:  For each host genotype, what is the overllap of DE genes over time points?\n",
    "\n",
    "Is there a great proportion of DEG shared across all time points? What are the time point combinations with the most shared DEG? What is the time point with the highest proportion of exclusive DEG?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the results of each time point and get the rownames, which contain the number of the gene.\n",
    "rh24_gi <- rownames(subset(rh24$res, rh24$res$padj <= 0.05))\n",
    "rh48_gi <- rownames(subset(rh48$res, rh48$res$padj <= 0.05))\n",
    "rh72_gi <- rownames(subset(rh72$res, rh72$res$padj <= 0.05))\n",
    "\n",
    "ven_rh <- list(\n",
    "    rh24=rh24_gi,\n",
    "    rh48=rh48_gi,\n",
    "    rh72=rh72_gi\n",
    ")\n",
    "\n",
    "rh_ven <- venn.diagram(ven_rh, main=\"Resistant DE over time\", filename=NULL,print.mode=c(\"raw\", \"percent\"))\n",
    "\n",
    "# Subset the results of each time point and get the rownames, which contain the number of the gene.\n",
    "sh24_gi <- rownames(subset(sh24$res, sh24$res$padj <= 0.05))\n",
    "sh48_gi <- rownames(subset(sh48$res, sh48$res$padj <= 0.05))\n",
    "sh72_gi <- rownames(subset(sh72$res, sh72$res$padj <= 0.05))\n",
    "\n",
    "ven_sh <- list(\n",
    "    sh24=sh24_gi,\n",
    "    sh48=sh48_gi,\n",
    "    sh72=sh72_gi\n",
    ")\n",
    "\n",
    "sh_ven <- venn.diagram(ven_sh, main=\"Susceptible DE over time\", filename=NULL, print.mode=c(\"raw\", \"percent\"))\n",
    "\n",
    "library(repr)\n",
    "options(repr.plot.height=4)\n",
    "grid.arrange(gTree(children=rh_ven), gTree(children=sh_ven), ncol=2, nrow=1, heights=c(1), widths=c(1,1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overall, the overlap of DEG for each phenotype across all combinations of time points was fairly low. The number of  DEG significant at **all** time points was 45 (6.23%) and 10 (2.38%) for the resistant and susceptible phenotypes, respectively. \n",
    "\n",
    "Starting with 24HPI point, we can see that the majority of the DEG in the **resistant** phenotype (67%) are actually shared with other time points. In fact, almost half (44.11%) of the DEG genes at 24HPI remain differentially expressed across all remaining time points. However, genes that are DE at 24h very rarely (3.9%) remain expressed only until the 48HPI point. On the other hand, 19.61% of the genes with differential expression at 24HPI cease to have differential expression at 48h and return as DEG at the 72HPI point. This trend is very similar to the one observed for the **susceptible** phenotype.\n",
    "\n",
    "At the 48HPI point, we can see that in both phenotyes, the majority of the DGE (59.34% for resistant and 57.20% forsusceptible) are shared with the 72HPI. As previously noted, there are very few DGE shared only with the 24HPI point.\n",
    "\n",
    "At the 72HPI point, we can see that (particularly for the resistant phenotype) the majority of the DEG are exclusive of that time point. \n",
    "\n",
    "The overall conclusion from these results are that there is tendency for genes that are differentially expressed at a given time point, to continue to be so along time. Nevertheless, the proportion of DEG that occur at all time points is very low. It also seems to be very rare for genes to be differentially expressed at the first two time points and then cease to be so in the 72HPI point. The majority of the DEG seem to appear at the last time point, as demonstrated by the fact that the number of exclusive genes for the 72HPI is always the highest."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2.2 Question:  For each host time point, what is the overlap of DE genes between phenotypes?\n",
    "\n",
    "Are the same genes being expressed in both phenotypes at each time point? And is this sharing consistent across time points? At which time point do they share the most DEG?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the results of each time point and get the rownames, which contain the number of the gene.\n",
    "rh24_gi <- rownames(subset(rh24$res, rh24$res$padj <= 0.05))\n",
    "sh24_gi <- rownames(subset(sh24$res, sh24$res$padj <= 0.05))\n",
    "\n",
    "ven_24h <- list(\n",
    "    rh24=rh24_gi,\n",
    "    sh24=sh24_gi\n",
    ")\n",
    "\n",
    "h24_ven <- venn.diagram(ven_24h, main=\"24HPI overlap between phenotypes\", filename=NULL, print.mode=c(\"raw\", \"percent\"))\n",
    "\n",
    "# Subset the results of each time point and get the rownames, which contain the number of the gene.\n",
    "rh48_gi <- rownames(subset(rh48$res, rh48$res$padj <= 0.05))\n",
    "sh48_gi <- rownames(subset(sh48$res, sh48$res$padj <= 0.05))\n",
    "\n",
    "ven_48h <- list(\n",
    "    rh48=rh48_gi,\n",
    "    sh48=sh48_gi\n",
    ")\n",
    "\n",
    "h48_ven <- venn.diagram(ven_48h, main=\"24HPI overlap between phenotypes\", filename=NULL, print.mode=c(\"raw\", \"percent\"))\n",
    "\n",
    "# Subset the results of each time point and get the rownames, which contain the number of the gene.\n",
    "rh72_gi <- rownames(subset(rh72$res, rh72$res$padj <= 0.05))\n",
    "sh72_gi <- rownames(subset(sh72$res, sh72$res$padj <= 0.05))\n",
    "\n",
    "ven_72h <- list(\n",
    "    rh72=rh72_gi,\n",
    "    sh72=sh72_gi\n",
    ")\n",
    "\n",
    "h72_ven <- venn.diagram(ven_72h, main=\"72HPI overlap between phenotypes\", filename=NULL, imagetype=\"png\",print.mode=c(\"raw\", \"percent\"))\n",
    "\n",
    "library(repr)\n",
    "options(repr.plot.height=6)\n",
    "blank<-rectGrob(gp=gpar(col=\"white\"))\n",
    "grid.arrange(gTree(children=h24_ven), blank, gTree(children=h48_ven), gTree(children=h72_ven), ncol=3, nrow=2, widths=c(0.8, 0.1, 0.8))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assessing the overlap between host phenotypes across all time points provides an interesting insight. Despite starting with rather different sets of DEG (8 genes; 27%), there is a clear trend for the increase in the overlap of DEG over time. This increases reaches its maximum at the last time point where almost 80% of the DEG in the susceptible phenotype are also found in the set of DEG from the resistant phenotype.\n",
    "\n",
    "A tantalizing suggestion from these results, is that the core response for \"resistance\" may not be very different between both genotypes over time. However, the response is elicited much earlier in the resistant phenotype, and having an earlier kick-off to that response appears to be the key. Over time, the overlap of DEG for both hosts clearly seems to converge in a pattern similar to an eclipse."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 Question: How does the fold change distributes according to the normalized read counts for each condition?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can also plot the distribution of the log2 fold changes for each condition over the mean of normalized counts for all samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "multiMA <- function(x) {\n",
    "    plotMA(x$res)\n",
    "}\n",
    "\n",
    "par(mfrow = c(2,3), mar=c(7, 4, 3, 0))\n",
    "\n",
    "\n",
    "y <- lapply(pairsr, multiMA)\n",
    "mtext(\"Resistant\", side = 3, line = -2, outer = TRUE)\n",
    "mtext(\"Susceptible\", side = 3, line = -28, outer = TRUE)\n",
    "mtext(\"24HPI\", at = 0.18, side = 3, line = -52, outer = TRUE)\n",
    "mtext(\"48HPI\", at = 0.53, side = 3, line = -52, outer = TRUE)\n",
    "mtext(\"72HPI\", at = 0.85, side = 3, line = -52, outer = TRUE)\n",
    "y <- lapply(pairss, multiMA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plotFC <- function(ds, title) {\n",
    "    ds$chr <- mcols(gdds)$chr;\n",
    "    mdf <- data.frame(\"gene\" = rownames(ds),\n",
    "                      \"chr\" = mcols(gdds)$chr ,\n",
    "                      \"fc\" = ds$log2FoldChange,\n",
    "                      \"qval\" = ds$padj)\n",
    "    df <- subset(mdf, mdf$qval <= 0.05)\n",
    "    dfc <- count(df, \"chr\")\n",
    "    p <- ggplot(df, aes(as.numeric(rownames(df)), fc)) +\n",
    "        geom_point(aes(color=fc)) +\n",
    "        geom_hline(data = dfc,\n",
    "                   aes(yintercept=freq * (max(df$fc) / max(dfc$freq)), alpha=.7, text=paste(\"Frequency: \", freq)),\n",
    "                   colour=\"red\") + \n",
    "        facet_grid(~chr, scales=\"free_x\") +\n",
    "        scale_colour_gradientn(colours=rainbow(7)) +\n",
    "        xlab(\"Genome position\") +\n",
    "        ggtitle(paste(title, \"(\", sum(dfc$freq) ,\")\")) + \n",
    "        theme(axis.text.x=element_blank(),\n",
    "              axis.ticks.x=element_blank(),\n",
    "              panel.spacing=unit(0.15, \"lines\"))\n",
    "        #scale_y_continuous(name = \"log2(FC)\",\n",
    "        #                   sec.axis = sec_axis(~ . * (max(df$fc) / max(dfc$freq)) , name = \"Frequency\"),\n",
    "        #                   limits = c(min(df$fc), max(df$fc)))\n",
    "    \n",
    "    p\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The number of DE genes tends to increase with the number of HPI and with a clear distribution skewed torwards up-regulated genes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Question: How do DEG distribute for each condition across the genome?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resistant phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plotFC(rh24$res, \"Resistant 24HPI\")\n",
    "plotFC(rh48$res, \"Resistant 48HPI\")\n",
    "plotFC(rh72$res, \"Resistant 72HPI\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Susceptible phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plotFC(sh24$res, \"Susceptible 24HPI\")\n",
    "plotFC(sh48$res, \"Susceptible 48HPI\")\n",
    "plotFC(sh72$res, \"Susceptible 72HPI\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Time-course DE\n",
    "\n",
    "For a control and treatment time series, one can use a design formula containing the condition factor, the time factor, and the interaction of the two. In this case, using the likelihood ratio test with a reduced model which does not contain the interaction terms will test whether the condition induces a change in gene expression at any time point after the reference level time point (time 0).\n",
    "\n",
    "Time-course experiments allow the explicit discovery of genes that have differential expression between two conditions **over time**. Previously, we tested whether a particular gene was differentially expressed between host phenotypes at any particular point in time. In this experiment, DE genes with low p-values will be the ones that show a host-specific effect **over time** after time 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "get_tcstats <- function(ds) {\n",
    "    design(ds) <- formula(~ cond + hpi + cond:hpi);\n",
    "    dsd <- DESeq(ds, test=\"LRT\", reduced=~ cond + hpi);\n",
    "    res <- results(dsd);\n",
    "    return(list(\"res\"=res, \"ds\"=dsd))\n",
    "}\n",
    "\n",
    "get_tcstats_cond <- function(ds) {\n",
    "    design(ds) <- formula(~ exp + hpi + exp:hpi);\n",
    "    dsd <- DESeq(ds, test=\"LRT\", reduced=~ exp + hpi);\n",
    "    res <- results(dsd, alpha=0.05);\n",
    "    return(list(\"res\"=res, \"ds\"=dsd))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3º Question: How does gene expression between control and inoculated conditions evolve over time?\n",
    "\n",
    "### 3.1 Question: Which genes show DE over time for the resistant phenotype? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the data to include only Q samples\n",
    "rqt = colData(gdds)$exp==\"R\"\n",
    "rds <- subset(gdds, select=rqt)\n",
    "\n",
    "rtc <- get_tcstats(rds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(which(rtc$res$padj <= 0.05))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(rtc$res), file=\"rtc.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rtc_ids = row.names(as.data.frame(subset(rtc$res, rtc$res$padj <= 0.05)))\n",
    "write(rtc_ids, \"resistant_timecourse.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the resistant phenotype, there are 83 genes that show a significant change in expression between control and inoculated conditions over time. This means that, at some point in time, these genes displayed a different expression over the transition to a particular time point between the control and inoculated groups.\n",
    "\n",
    "We can also plot the profile of each DEG over time using an heatmap as below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "betas <- coef(rtc$ds)\n",
    "topGenes <- head(order(rtc$res$padj), 83)\n",
    "\n",
    "mat <- betas[topGenes, -c(1,2)]\n",
    "\n",
    "thr <- 3\n",
    "mat[mat < -thr] <- -thr\n",
    "mat[mat > thr] <- thr\n",
    "\n",
    "pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101), cluster_col=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the heatmap above, we can see from the last 2 columns (condQ.hpi48h and condQ.hpi72h) the expression profile of each of the 83 significant genes over time. For example, the first gene (row) displayed a strong induction of expression between the 24-48h and then regressed to expression levels similar to the control. For the majority of these genes, there seems to be a strong induction of expression over time, reaching maximum values when transitioning to the last time point. However, a group of roughly 20 genes (at the bottom of the heatmap) also displayed a strong downregulation on the inoculated group over time.\n",
    "\n",
    "If we were interested in the particular profile of a certain gene, we could also model the read counts of that gene as a smooth function of time, and include an interaction term of the condition with the smooth function. Let's demonstrate with the first gene in the above heatmap, 18841."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "sprof <- function(gnID) {\n",
    "    fiss <- plotCounts(rtc$ds, gnID, \n",
    "                       intgroup = c(\"hpi\",\"cond\"), returnData = TRUE)\n",
    "    p <- ggplot(fiss,\n",
    "      aes(x = as.numeric(hpi), y = count, color = cond, group = cond)) +\n",
    "      geom_point() + geom_smooth(se = FALSE, method = \"loess\") + scale_y_log10()\n",
    "    return(p)\n",
    "}\n",
    "\n",
    "gene27 <- head(order(rtc$res$padj), 27)\n",
    "\n",
    "z <- lapply(gene27, sprof)\n",
    "\n",
    "options(repr.plot.height=16)\n",
    "multiplot(plotlist=z, cols=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Question: Which genes show DE over time for the resistant phenotype? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the data to include only Q samples\n",
    "sqt = colData(gdds)$exp==\"S\"\n",
    "sds <- subset(gdds, select=sqt)\n",
    "\n",
    "stc <- get_tcstats(sds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(which(stc$res$padj <= 0.05))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "write.table(as.matrix(stc$res), file=\"stc.csv\", , quote=F, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "stc_ids = row.names(as.data.frame(subset(stc$res, stc$res$padj <= 0.05)))\n",
    "write(stc_ids, \"susceptible_timecourse.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, the number DEG over time for the susceptible host is much lower than in the resistant host. This is an expected result from the previous analyses where each time point was compared independently for both hosts, where the susceptible host showed less DEG.\n",
    "\n",
    "We can perform the same heatmap analysis to see the profile of the significant DE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sbetas <- coef(stc$ds)\n",
    "stopGenes <- head(order(stc$res$padj), 25)\n",
    "\n",
    "smat <- betas[stopGenes, -c(1,2)]\n",
    "\n",
    "thr <- 3\n",
    "smat[smat < -thr] <- -thr\n",
    "smat[smat > thr] <- thr\n",
    "\n",
    "options(repr.plot.height=6)\n",
    "pheatmap(smat, breaks=seq(from=-thr, to=thr, length=101), cluster_col=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unlike the heatmap for the resistant host, this showed a much lesser trend of strong induction in the inoculated group. In fact, almost half of the DEG showed an variable degree of down-regulation, compared to the control group.\n",
    "\n",
    "We can now assess the overlap of these DEG between the resistant and susceptible hosts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "rde <- which(rtc$res$padj <= 0.05)\n",
    "sde <- which(stc$res$padj <= 0.05)\n",
    "\n",
    "tc_ds <- list(\n",
    "    \"rv\" = rde,\n",
    "    \"sv\" = sde\n",
    ")\n",
    "options(repr.plot.height=5)\n",
    "tc_ven <- venn.diagram(tc_ds, filename=NULL,  imagetype=\"png\",print.mode=c(\"raw\", \"percent\"))\n",
    "grid.draw(tc_ven)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interestingly, only one DEG is shared between the resistant and susceptible hosts, which means that the genes that are being elicited over time are almost completely different between these hosts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we can also plot the distribution of the DEG over time along the genome for both hosts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotFC(rtc$res, \"DEG over time for Resistant\")\n",
    "plotFC(stc$res, \"DEG over time for Susceptible\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is noteworthy that, as in the independent time point comparisons performed before, Chr2 consistently displays the highest number of DEG."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4º Question: Which genes are DE at each time point between control and inoculated in function of the host phenotype?\n",
    "\n",
    "The question that will be addressed here is which genes are DE between control vs. inoculated **and** resistant vs. susceptible. The rationale is very similar to the time-course experiment, but replacing the cond:hpi interaction with cond:exp."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hpi24_c <- colData(gdds)$hpi==\"24h\"\n",
    "hpi48_c <- colData(gdds)$hpi==\"48h\"\n",
    "hpi72_c <- colData(gdds)$hpi==\"72h\"\n",
    "\n",
    "hpi_intest <- function(sel) {\n",
    "    ss <- subset(gdds, select=sel)\n",
    "    design(ss) <- formula(~ cond + exp + cond:exp);\n",
    "    dsd <- DESeq(ss, test=\"LRT\", reduced=~ cond + exp);\n",
    "    res <- results(dsd, alpha=.05);\n",
    "    print(length(which(res$padj <= 0.05)))\n",
    "    return(list(\"res\"=res, \"ds\"=dsd))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpi24i <- hpi_intest(hpi24_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpi48i <- hpi_intest(hpi48_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hpi72i <- hpi_intest(hpi72_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "na_conv <- function(df) {\n",
    "    ndf <- df;\n",
    "    ndf[ndf > 0.05] <- NA;\n",
    "    return(ndf)\n",
    "}\n",
    "\n",
    "ds_v <- list(rh24$res$padj,rh48$res$padj, rh72$res$padj,\n",
    "             sh24$res$padj, sh48$res$padj, sh72$res$padj,\n",
    "             rtc$res$padj, stc$res$padj,\n",
    "             hpi24i$res$padj, hpi48i$res$padj, hpi72i$res$padj)\n",
    "\n",
    "ds_p <- rapply(ds_v, na_conv, how=\"list\")\n",
    "\n",
    "main_df <- data.frame(\n",
    "    \"rh24\"=ds_p[[1]],\n",
    "    \"rh48\"=ds_p[[2]],\n",
    "    \"rh72\"=ds_p[[3]],\n",
    "    \"sh24\"=ds_p[[4]],\n",
    "    \"sh48\"=ds_p[[5]],\n",
    "    \"sh72\"=ds_p[[6]],\n",
    "    \"rtc\"=ds_p[[7]],\n",
    "    \"stc\"=ds_p[[8]],\n",
    "    \"hpi24i\"=ds_p[[9]],\n",
    "    \"hpi48i\"=ds_p[[10]],\n",
    "    \"hpi72i\"=ds_p[[11]]\n",
    ")\n",
    "\n",
    "#Remove rows containing only NA\n",
    "final_df <- main_df[rowSums(is.na(main_df)) != 11,]\n",
    "\n",
    "nrow(final_df)\n",
    "\n",
    "write.csv(final_df, file=\"DEG.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3º Question: Which genes are DEG in a control setting between the two hosts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c24 <- getpair_stats(gdds, \"c24\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c48 <- getpair_stats(gdds, \"c48\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c72 <- getpair_stats(gdds, \"c72\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotFC(c24$res, \"este\")\n",
    "plotFC(c48$res, \"este\")\n",
    "plotFC(c72$res, \"este\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
